<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>The Database-Integrated Genome Screening (DIGS) Tool by giffordlabcvr</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">

  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-805749-23', 'auto');
    ga('send', 'pageview');

  </script>

  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/tarball/master" class="btn">Download .tar.gz</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
    </section>

    <section class="main-content">





<h3>
<a id="database-integrated-genome-screening-digs" class="anchor" href="#database-integrated-genome-screening-digs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Database-integrated genome screening (DIGS)</strong>
</h3>


<p>
Molecular sequence data are highly information rich, and are now being generated much faster than they can be analysed. Consequently, public databases contain multitudes of gene, <a href="http://pseudogene.org/background.php">pseudogene</a>, <a href="http://www.nature.com/scitable/topicpage/transposons-the-jumping-genes-518">transposon</a>, virus, and <a href="https://en.wikipedia.org/wiki/Noncoding_DNA">non-coding element</a> sequences
 that have not been yet been identified, or are only poorly described.
</p>


<p>
<strong>Database-integrated genome screening (DIGS) can be used to screen genome and transcriptome assemblies for sequence features of interest (e.g. genes, transposons, functional non-coding sequences), facilitating investigations of their evolution, distribution and diversity.</strong> 
</p>


<p>
In DIGS, the output of <a href="http://www.ebi.ac.uk/Tools/sss/">sequence similarity</a> search-based screens is captured in a <a href="https://docs.oracle.com/javase/tutorial/jdbc/overview/database.html">relational database</a>.
This allows for the interrogation and manipulation of output data using <a href="http://www.w3schools.com/sql/sql_intro.asp">structured query language (SQL)</a>.
In addition, it provides all the benefits of a relational database management system (RDBMS) with respect to features such as data recoverability, multi-user support and network access.
</p>




<h3>
<a id="the-digs-tool" class="anchor" href="#the-digs-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>The DIGS tool</strong>
</h3>


<p>The <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki">DIGS tool</a> is a <a href="https://www.perl.org/">PERL</a> program for implementing DIGS with assembled sequence data (not short read data). 
It uses the <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">Basic Local Alignment Search Tool (BLAST)</a> to perform sequence similarity searches, and the <a href="https://www.mysql.com/">MySQL</a> RDBMS to capture their output.
</p>


<p>
Instructions for installing and running the DIGS tool can be found on <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki">the DIGS tool wiki</a>
</p>




<h3>
<a id="digs-projects" class="anchor" href="#digs-projects" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Projects using DIGS</strong>
</h3>

<p>
<a href="./digs-for-eves.html">DIGS for EVEs</a>: Using DIGS to investigate the virus "fossil record".
</p>





<h3>
	<a id="author" class="anchor" href="#author" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Author</strong>
</h3>

<p>Robert J. Gifford (<a href="mailto:robert.gifford@glasgow.ac.uk">robert.gifford@glasgow.ac.uk</a>)</p>




      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">The Database-Integrated Genome Screening (DIGS) Tool</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>


  </body>
</html>
