<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS Tool</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="website/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="website/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
    <section class="page-header">
    
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">


	  <h3>
		<a id="the-digs-tool" class="anchor" href="#the-digs-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>The DIGS tool</strong>
	  </h3>
      <hr>

		<p>
		  The
		  <strong>database-integrated genome-screening (DIGS) tool</strong>
		  is an open source
		  <a target="_blank" href="https://www.perl.org/"><strong>PERL</strong></a>
		  program for implementing systematic <i>in silico</i> <strong>genome screening</strong>.
		</p>

	  <h3>
	  <a id="genome-screening" class="anchor" href="#database-integrated-genome-screening-digs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>What is <i>in silico</i> genome-screening?</strong>
	  </h3>
      <hr>


		<p>
		  Molecular sequence data are highly information rich, and are now being generated
		  much faster than they can be analysed. Consequently, published genome sequences
		  contain many features (e.g. genes, pseudogenes, transposons) that have not been yet
		  been formally described, or are unreliably annotated.
		</p>


		<p>
         <b>Sequence similarity search tools</b>, such as those implemented in the
         <a target="_blank" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">
		 <strong>Basic Local Alignment Search Tool (BLAST)</strong></a>
         can be used to systematically screen
         whole genome data and recover sets of related sequences
         <b>without relying on previously generated genome annotations</b>. 
		</p>


		<p>

         This approach can form the basis for comparative studies focussing on specific
         genome features such as genes, mobile genetic elements, or conserved protein
         domains. In addition, similarity searches can in themselves provide
         useful insights into the distribution of conserved sequence features, for example
         by revealing differences in gene copy number between species.
		</p>



	  <h3>
	  <a id="database-integrated-genome-screening-digs" class="anchor" href="#database-integrated-genome-screening-digs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>What is database-integrated genome-screening (DIGS)</strong>
	  </h3>
      <hr>


	  <p>
		In <b>database-integrated genome-screening (DIGS)</b>, the output of sequence similarity
		search-based screens is captured in a relational database
		This allows for the interrogation and manipulation of output data using
		structured query language (SQL).
	  </p>


	  <h3>
		<a id="requirements" class="anchor" href="#the-digs-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Requirements</strong>
	  </h3>
      <hr>


	  <p>
		<strong>Similarity searches</strong>: the DIGS tool uses the
		<a target="_blank" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">
		<strong>Basic Local Alignment Search Tool (BLAST)</strong></a>
		to perform sequence similarity searches. 
	  </p>

	  <p>
		<strong>Relational Database</strong>: the DIGS tool uses the
		<a target="_blank" href="https://www.mysql.com/"><strong>MySQL</strong></a>
		relational database management system (RDBMS) to coordinate  <i>in silico</i>
		screens and to capture their output.

	  </p>


		Instructions for installing and running the DIGS tool can be found on
		<a href="https://github.com/giffordlabcvr/DIGS-tool/wiki">the DIGS tool wiki</a>
	  </p>




	  <h3>
	    <a id="author" class="anchor" href="#author" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Author</strong>
	  </h3>
      <hr>


	  <p>Robert J. Gifford (<a href="mailto:robert.gifford@glasgow.ac.uk">robert.gifford@glasgow.ac.uk</a>)</p>


      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">The Database-Integrated Genome Screening (DIGS) Tool</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  </body>

</html>
