<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS Tool</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="website/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="website/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
    <section class="page-header">
    
      <h1 class="project-name">DIGS Tool</h1>
      <h2 class="project-tagline">Database-Integrated Genome Screening (DIGS).</h2>
      
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">


	  <h3>
		<a id="the-digs-tool" class="anchor" href="#the-digs-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Explore the dark genome using BLAST and a relational database</strong>
	  </h3>
      <hr>

		<p>
		  The
		  <strong>database-integrated genome-screening (DIGS) tool</strong>
		  is an open source software framework
		  for implementing '<strong><i>in silico</i> genome screening</strong>'.
		  This entails using sequence similarity search tools to 
		  systematically search molecular sequence databanks for groups of related sequences,
		  usually  with a view to better understanding their evolution, and thereby gaining 
		  insights into their biology.
		</p>

	  
		  
		<p>
		  The DIGS tool allows users to leverage the combined power of
		  <b>sequence similarity searches</b> and <b>structured query language (SQL)</b>
		  to investigate the <b>distribution, diversity and evolution</b> of conserved genome
		  features, including those that have not yet been recognised or annotated.
		</p>
		<br>



      <p><img src="./website/images/digs-screenshot.png" alt="DIGS screenshot" /></p>
          

	  <blockquote>
		  <b>Database integrated genome screening</b>: this screenshot shows 
		  querying of a DIGS project database. 
		  In this example, the
          <a target="_blank" href="https://www.sequelpro.com/"><strong>SequelPro</strong></a>
		  SQL client has been used to connect.
		  The screening database shown here contains the results of a broad screen
		  across mammalian species to identify rare 'endogenous lentivirus'
		  sequences, such as 'rabbit endogenous lentivirus K' 
          (<a target="_blank" href="https://doi.org/10.1073/pnas.0700471104"><strong>RELIK</strong></a>).           
	  </blockquote>



        
        
	   <br>


		
		

	  <h3>
	  <a id="database-integrated-genome-screening-digs" class="anchor" href="#database-integrated-genome-screening-digs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Background</strong>
	  </h3>
      <hr>


		<p>
		  Molecular sequence data are highly information rich, and are now being generated
		  much faster than they can be analysed. Consequently, published genome sequences
		  contain many features (e.g. genes, pseudogenes, transposons) that have not been yet
		  been formally described, or are unreliably annotated.
         However, <b>sequence similarity searches</b> provide a means to 
         recover sets of related sequences from databanks
         without relying on previously generated annotations. 
		</p>




	  <h3>
		<a id="requirements" class="anchor" href="#the-digs-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Requirements</strong>
	  </h3>
      <hr>


	  <p>
		<strong>Similarity searches</strong>: the DIGS tool uses the
		<a target="_blank" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">
		<strong>Basic Local Alignment Search Tool (BLAST)</strong></a>
		to perform sequence similarity searches. 
	  </p>

	  <p>
		<strong>Relational Database</strong>: the DIGS tool uses the
		<a target="_blank" href="https://www.mysql.com/"><strong>MySQL</strong></a>
		relational database management system (RDBMS) to coordinate  <i>in silico</i>
		screens and to capture their output.

	  </p>


		Instructions for installing and running the DIGS tool can be found on
		<a href="https://github.com/giffordlabcvr/DIGS-tool/wiki">the DIGS tool wiki</a>
	  </p>




	  <h3>
	    <a id="author" class="anchor" href="#author" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Author</strong>
	  </h3>
      <hr>


	  <p>Robert J. Gifford (<a href="mailto:robert.gifford@glasgow.ac.uk">robert.gifford@glasgow.ac.uk</a>)</p>


      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">The Database-Integrated Genome Screening (DIGS) Tool</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  </body>

</html>
