[[images/UserGuideCover.png]]


The DIGS tool provides a computational framework for implementing **database-integrated genome-screening (DIGS)**. The DIGS tool is written in [PERL](https://www.perl.org/), it uses  the [Basic Local Alignment Search Tool (BLAST)](http://blast.ncbi.nlm.nih.gov/Blast.cgi) to perform to perform sequence similarity searches, and [MySQL](https://www.mysql.com/) as an RDBMS.

We originally created the DIGS tool for research in the area of ‘[paleovirology](http://bioinformatics.cvr.ac.uk/paleovirology/site/html/posts/2013-04-15_what_is_paleovirology.html)’, and have primarily used it to screen genome sequence databases for sequences derived from viruses. However, we expect it may have broader utility, and it has been developed with aim of saving others the effort of creating a similar system. 



