<!DOCTYPE html>
<html lang="en-us">

  <head>

	<meta charset="UTF-8">
	<title>DIGS for EVEs</title>
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
	<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet-eve.css" media="screen">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">
	
  </head>
  
  <body>
  
    <section class="page-header">
      <h1 class="project-name">DIGS for EVEs</h1>
      <h2 class="project-tagline">Database-integrated genome screening (DIGS) for endogenous viral elements (EVEs)</h2>
      <a href="../../index.html" class="btn">DIGS Home</a>
      <a target="_blank" href="http://glue-tools.cvr.gla.ac.uk" class="btn">GLUE Home</a>
      <a href="https://twitter.com/DIGS_for_EVEs" class="btn">Twitter feed</a>
      <!--<a href="./eve.txt" class="btn">EVE catalog</a>-->
    </section>

    <section class="main-content">

	   <h3>
		   <a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Overview</strong>
	   </h3>
	   <hr>
	  

		 <p>
		 This is the home page of <strong>DIGS for EVEs</strong>, 
		 a study lead by <a target="_blank" href="https://giffordlabcvr.github.io/virus-evolution/">Robert J. Gifford</a>.
		 </p> 


		 <p><img src="../assets/images/digs-for-eves-header.jpg" alt="DIGS-for-EVEs header" /></p>


		 <p>
		   In this project I have used the
		   <a href="http://giffordlabcvr.github.io/DIGS-tool/"><strong>database integrated genome screening (DIGS) tool</strong></a>
		   to systematically screen metazoan genomes <i>in silico</i> with the aim of recovering
		   <strong>viral 'fossil record'</strong>. This record is comprised of 
		   <strong>endogenous viral elements (EVEs)</strong> - DNA sequences found in eukaryotic genomes that 
		   derive from ancient viruses.
		 </p> 


		 <p>
		   I've attempted to make the viral fossil record more readily accessible to 
		   interested researchers through the: (i) introduction of a standard nomenclature,
		   and; (ii) the use of the <a target="_blank" href="http://glue-tools.glue.cvr.ac.uk/"><b>GLUE</b></a>
		   software framework to capture the complex semantic links among data items. 
		 </p> 


	   <h3>
		 <a id="scopeTimeframe" class="anchor" href="#scopeTimeframe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Scope</strong>
	   </h3>
	   <hr>
	  

		 <p>
		 The DIGS for EVEs project is currently restricted to EVEs derived from viruses
		 with genome sizes <strong>under 30 kilobases</strong>. Thus, the project excludes 
		 poxviruses and other double-stranded DNA viruses with similarly large genomes.
		 </p>
	   
		 <p>
		 DIGS for EVEs is an open-ended project for several reasons. Firstly, new genome data are continually accumulating,
		 and every new genome offers the possibility to identify novel EVEs.
		 </p>

		 <p>
		 Secondly, all EVEs have been detected based on their similarity in sequence to viruses. Characterisation of novel viruses can therefore 
		 lead to the identification of new EVEs in published genomes.
		 </p>

		 <p>
		 Finally, discriminating EVEs from other kinds of genomic sequences is straightforward for some virus groups,
		 but less so for others. In challenging cases, carrying out the kinds of detailed investigation required to positively identify EVEs
		 is a slow process.
		 </p>

		 <p>
		 In this project, we aim to focus on identifying the most tractable kinds of EVE, across the broadest possible
		 range of Eukaryotic hosts. Scope is expected to expand incrementally as the project progresses.
		 </p>


	   <h3>
		 <a id="glueResources" class="anchor" href="#glueResources" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>GLUE projects</strong>
	   </h3>
	   
	   <p>
         The GLUE software framework aims to enable reuse of virus genome data and
         associated algorithms across different sequence analysis contexts. 



		 We have incorporated data recovered in the DIGS-for-EVEs project into 
		 'sequence-based resources' constructed using <a target="_blank" href="http://glue-tools.glue.cvr.ac.uk/"><b>GLUE</b></a>.
	   </p>
	   
	   <p>
	     <a target="_blank" href="https://giffordlabcvr.github.io/Hepadnaviridae-GLUE/"><strong>Hepadnaviridae-GLUE</strong></a>:
	     a hepadnavirus-focussed project that captures information about <strong>hepadnaviruses</strong> 
	     (<a target="_blank" href ="https://viralzone.expasy.org/9?outline=all_by_species">
         family <i>Hepadnaviridae</i></a>) -a group of reverse-transcribing DNA viruses that
         infect vertebrates. The type species - <b>hepatitis B virus (HBV)</b> - is estimated to
         infect ~300 million people worldwide, causing substantial morbidity and mortality.
         Recent studies have revealed that hepadnaviruses infect a diverse range of
         vertebrate species, ranging from fish to mammals. They are associated with
         disease in many of these species. 
	   </p>
	   
	   <p>
	     <a target="_blank" href="https://giffordlabcvr.github.io/CRESS-GLUE/"><strong>CRESS-GLUE</strong></a>:
	     nothing to do with sandwiches - this GLUE repository focuses on circular Rep-encoding single-stranded DNA
		 (<a target="_blank" href="https://www.sciencedirect.com/science/article/pii/S0065352718300575"><b>CRESS DNA</b></a>)
		 viruses (phylum 
		 <a target="_blank" href="https://jvi.asm.org/content/94/12/e00582-20"><b><i>Cressdnaviricota</i></b>)</a>.
	   </p>
	   <p><a target="_blank" href="https://giffordlabcvr.github.io/Parvovirus-GLUE/"><strong>Parvovirus-GLUE</strong></a>: [pending release 2021]</p>
	   <p><a target="_blank" href="https://giffordlabcvr.github.io/Flavivirus-GLUE/"><strong>Flavivirus-GLUE</strong></a>: Resources for comparative genomic analysis of flaviviruses.

		 The flaviviruses (family <i>Flaviviridae</i>) are a group of enveloped, positive-strand
		 RNA viruses,
		 many of which pose serious risks to human health on a global scale.
		 Arthropod-borne flaviviruses such as <b>Zika virus (ZIKV)</b>, <b>Dengue virus (DENV)</b>,
		 and yellow fever virus (YFV) are the causative agents of large-scale outbreaks that
		 result in millions of human infections every year, while the bloodborne hepatitis C
		 virus (HCV) is a major cause of chronic liver disease.
 	   </p>
       


	   <p><a target="_blank" href="https://giffordlabcvr.github.io/Filovirus-GLUE/"><strong>Filovirus-GLUE</strong></a>: [pending release 2021]</p>
	   <!--<p><a target="_blank" href="https://giffordlabcvr.github.io/Bornavirus-GLUE/"><strong>Bornavirus-GLUE</strong></a></p>-->
	   <!--<p><a target="_blank" href="https://giffordlabcvr.github.io/Peribunyavirus-GLUE/"><strong>Peribunyavirus-GLUE</strong></a></p>-->



	   <p>
	     <a target="_blank" href="https://giffordlabcvr.github.io/HHV6-GLUE/"><strong>HHV6-GLUE</strong></a>:
	     human betaherpesvirus 6A (HHV-6A) and human betaherpesvirus 6B (HHV-6B) are closely 
	     related species of roseolovirus (genus <i>Roseolovirus</i>). These viruses cause
	     lifelong, persistent infections and are known to sometimes integrate into the telomeric
	     regions of chromosomes. Insertions into the germline sometimes occur, and a small
	     proportion of people (1-2%) carry chromosomally-integrated HHV6 (ciHHV6) insertions
	     in their genome.
	   </p>

	   <p>
	     <a target="_blank" href="https://giffordlabcvr.github.io/Deltaretrovirus-GLUE/"><strong>Deltaretrovirus-GLUE</strong></a>:
         deltaretroviruses are an unusual group of complex retroviruses, with only a few species known. This project
         contains all GenBank sequence data, as well as rare examples of deltaretrovirus-derived ERVs. 	     
	   </p>
	   
	   <p><a target="_blank" href="https://giffordlabcvr.github.io/Lentivirus-GLUE/"><strong>Lentivirus-GLUE</strong></a>: [pending release 2021]</p>
	   <p><a target="_blank" href="https://giffordlabcvr.github.io/ERVdb/"><strong>ERVdb</strong></a>: [pending release late 2022]</p>
	   


 	   <br>
	   <h3>
		 <a id="Nomenclature" class="anchor" href="#Nomenclature" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Nomenclature for ERVs and EVEs</strong>
	   </h3>
	   <hr>

		<p>
          We have applied a systematic approach to naming ERVs and EVEs,
         <a target="_blank" href="https://retrovirology.biomedcentral.com/articles/10.1186/s12977-018-0442-1"><b>described here</b></a>.
		  EVEs and ERVs are assigned a unique identifier (ID) constructed from a defined
		  set of components.
  		</p>
 
        <p><img src="../assets/images/nomenclature-example-ehbv.jpg" alt="EVE Nomenclature - Lenti example" /></p>


		<p>
          The first component is a classifier that usually denotes the virus family
          that the EVE derives from.
          For example, the classifier 'ERV' is applied to all endogenous retroviruses.
          Classifiers usually follow the conventions established previously, as far as possible.
          For example, endogenous hepadnaviruses are given the classifier eHBV (endogenous
          hepatitis B virus), as this term has generally been used to describe endogenous 
          hepadnaviruses.
          For segmented viruses classifiers also designate a gene or segment
          (e.g. EBLN = endogenous borna-like nucleoprotein).
		</p>

		<p>
          The second component is a composite of two distinct subcomponents separated by a period:
          
          (i)  the name of the specific subgroup it derives from;
          (ii) a numeric ID that uniquely identifies the insertion.
          The numeric ID is an integer that identifies a unique insertion locus that arose as a
          consequence of an initial germline infection. Thus, orthologous copies in different
          species are given the same number. 
 		</p>
		
		<p>
          Where an EVE sequence is thought to have been duplicated within the germline
          following it's initial incorporation (e.g. via segmental duplication or transposition)
          we have appended an additional 'duplicate id' to the numeric ID, separated by a period.
          Please note that we have not yet resolved the orthologous relationships among
          sets of eHBV sequences belonging to multicopy eHBV lineages. We
          have therefore assigned unique duplicate IDs to each sequence within these lineages. 
		</p>
		
		<p>
          The third component of the ID defines the set of host species in which
          the ortholog occurs, or did occur prior to being deleted.
  		</p>


	   <h3>
		 <a id="digsFiles" class="anchor" href="#digsFiles" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>DIGS screening files</strong>
	   </h3>
	   <hr>
	  
	 
	   <p>
	   Probe and reference sets of polypeptide sequences for DIGS were obtained from NCBI's
	   <a target="_blank" href="http://www.ncbi.nlm.nih.gov/genome/viruses/">'Viral Genomes</a> resource.
	   </p> 

	   <p>
		 In addition to reference sequences obtained from NCBI, the DIGS reference library included in this repo contains reference sequences for EVEs,
		 and for various non-EVE genomic sequences that give spurious matches to viruses in DIGS, including: (i) host genomic sequences; (ii) endogenous retroviruses; (iii) transposons. 
	   </p>
<!--
	   <p>
		 Download here: <a href="https://github.com/giffordlabcvr/EVE/zipball/master">zip file</a>, 
		 <a href="https://github.com/giffordlabcvr/EVE/tarball/master">tarball</a>
	   </p>
-->


	   <h3>
		 <a id="relatedResources" class="anchor" href="#relatedResources" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Related Resources</strong>
	   </h3>
	   <hr>
	  
	   <p>
		 <a  target="_blank" href="./virus-host-interface.html"><strong>Virus-Host Interface</strong></a>: 
         In these research projects, the
		 <a href="http://giffordlabcvr.github.io/DIGS-tool/"><strong>DIGS tool</strong></a>
         was used to investigate the evolution of virus-interacting proteins (VIPs) and the
         <a href="http://tools.glue.cvr.ac.uk/"><b>GLUE</b></a> software environment was used to
         capture domain knowledge recovered in each study.	 
	   </p>


	   <br>
	   <h3>
		 <a id="Publications" class="anchor" href="#Publications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Related Publications</strong>
	   </h3>
	   <hr>



  <p>
 

	   <p>
	   Bamford CGG, de Souza WM, Parry R and RJ Gifford
	   <b>(2021)</b>
	   <br>
	   Comparative analysis of genome-encoded viral sequences reveals the evolutionary history of the <i>Flaviviridae</i>.
	   <br>
	   preprint <a href="https://www.biorxiv.org/content/10.1101/2021.09.19.460981v1" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Lytras S, Arriagada G, and RJ Gifford
	   <b>(2021)</b>
	   <br>
	   Ancient evolution of hepadnaviral paleoviruses and their impact on host genomes.
	   <br>
	   <i>Virus Evolution</i>
	   <a href="https://academic.oup.com/ve/advance-article/doi/10.1093/ve/veab012/6157734" target="_blank">[view]</a>
	   <br>
	   <br>

	   Hildebrandt E, Penzes J, Gifford RJ, Agbandje-Mckenna M, and R Kotin
	   <b>(2020)</b>
	   <br>
	   Evolution of dependoparvoviruses across geological timescales – implications for design of AAV-based gene therapy vectors.
	   <i>Virus Evolution</i> 
	   <a href="https://academic.oup.com/ve/article/doi/10.1093/ve/veaa043/5842153">[view]</a> 
	   <br>
	   <br>

	   Pénzes JJ, de Souza WM, Agbandje-Mckenna M, and RJ Gifford
	   <b>(2019)</b>
	   <br>
	   An ancient lineage of highly divergent parvoviruses infects both vertebrate and invertebrate hosts.
	   <br>
	   <i>Viruses</i>
	   <a href="https://www.mdpi.com/1999-4915/11/6/525" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Callaway HM, Subramanian S, Urbina C, Barnard K, Dick R, Hafentein SL, Gifford RJ, and CR Parrish
	   <b>(2019)</b>
	   <br>
	   Examination and reconstruction of three ancient endogenous parvovirus capsid proteins in rodent genomes.
	   <br>
	   <i>Journal of Virology</i>
	   <a href="https://jvi.asm.org/content/93/6/e01542-18.long" target="_blank">[view]</a>
	   <br>
	   <br>

	   Hron T, Elleder D, and RJ Gifford
	   <b>(2019)</b>
	   <br>
	   Deltaretroviruses have circulated since at least the Paleogene and infected a broad range of mammalian species.
	   <br>
	   <i>Retrovirology</i>
	   <a href="https://retrovirology.biomedcentral.com/articles/10.1186/s12977-019-0495-9" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Halo JV, Pendleton AL, Jarosz AS, Gifford RJ, Day ML, and JM Kidd
	   <b>(2019)</b>
	   <br>
	   Origin and recent expansion of an endogenous gammaretroviral lineage in canids.
	   <br>
	   <i>Retrovirology</i>
	   <a href="https://retrovirology.biomedcentral.com/articles/10.1186/s12977-019-0468-z" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Zhu H, Gifford RJ*, and Murcia PR*
	   <b>(2018)</b>
	   <br>
	   Distribution, diversity and evolution of endogenous retroviruses in perissodactyl genomes.
	   <br>
	   <i>Journal of Virology</i>
	   <br>* co-corresponding authors
	   <a href="https://jvi.asm.org/content/92/23/e00927-18" target="_blank">[view]</a>
	   <br>
	   <br>

	   Zhu H, Dennis T, Hughes J, and RJ Gifford
	   <b>(2018)</b>
	   <br>
	   Database-integrated genome screening (DIGS): exploring genomes heuristically using sequence similarity search tools and a relational database.
	   <a href="https://www.biorxiv.org/content/early/2018/04/25/246835" target="_blank">[preprint]</a> 
	   <br>
	   <br>

	   Gifford RJ, Blomberg B, Coffin JM, Fan H, Heidmann T, Mayer J, Stoye J, Tristem M, and WE Johnson
	   <b>(2018)</b>
	   <br>
	   Nomenclature for endogenous retrovirus (ERV) loci.
	   <br>
	   <i>Retrovirology</i>
	   <a href="https://doi.org/10.1186/s12977-018-0442-1" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Pénzes JJ, Marsile-Medun S, Agbandje-McKenna M, and RJ Gifford
	   <b>(2018)</b>
	   <br>
	   Endogenous amdoparvovirus-related elements reveal insights into the biology and evolution of vertebrate parvoviruses.
	   <br>
	   <i>Virus Evolution</i>
	   <a href="https://academic.oup.com/ve/article/4/2/vey026/5179240" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Blanco Melo D, Gifford RJ, and P. Bieniasz
	   <b>(2018)</b>
	   <br>
	   Reconstruction of a replication-competent ancestral murine endogenous retrovirus-L.
	   <br>
	   <i>Retrovirology</i>
	   <a href="https://retrovirology.biomedcentral.com/articles/10.1186/s12977-018-0416-3" target="_blank">[view]</a>
	   <br>
	   <br>

	   Dennis TPW, Flynn PJ, de Souza WM, Singer JB, Moreau CS, Wilson SJ, and RJ Gifford 
	   <b>(2018)</b>
	   <br>
	   Insights into circovirus host range from the genomic fossil record.
	   <br>
	   <i>Journal of Virology</i>
	   <a href="http://jvi.asm.org/content/92/16/e00145-18.full" target="_blank">[view]</a> 
	   <br>
	   <br>

	   Dennis TPW, de Souza WM, Marsile-Medun S, Singer JB, Wilson SJ, and RJ Gifford 
	   <b>(2018)</b>
	   <br>
	   The evolution, distribution and diversity of endogenous circoviral elements in vertebrate genomes.
	   <br>
	   <i>Virus Research</i>
	   <a href="https://www.sciencedirect.com/science/article/pii/S0168170217309048" target="_blank">[view]</a> 
	   <br>
	   <br>
	   
	   Souza WM, Romeiro MF, Fumagalli MJ, Modha S, de Araujo J, Queiroz LH, Durigon EL, Figueiredo LT, Murcia PR, Gifford RJ.
	   <b>(2017)</b>
	   <br>
	   Chapparvoviruses occur in at least three vertebrate classes and have a broad biogeographic distribution.
	   <br>
	   <i>J Gen Virol.</i>
	   <a href="http://jgv.microbiologyresearch.org/content/journal/jgv/10.1099/jgv.0.000671" target="_blank">[view]</a>
	   <br>
	   <br>


	   Blanco-Melo D, Gifford RJ, and PD Bieniasz
	   <b>(2017)</b>
	   <br>
	   Co-option of an endogenous retrovirus envelope for host defense in hominid ancestors.
	   <br>
	   <i>Elife</i>
	   <a href="https://elifesciences.org/articles/22519" target="_blank">[view]</a>
	   <br>
	   <br>


   </p>


	   <br>
	   <h3>
		 <a id="license" class="anchor" href="#license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>License</strong>
	   </h3>
       <hr>

		 <p>
		   This project is licensed under the <a href="https://www.gnu.org/licenses/agpl-3.0.en.html">GNU Affero General Public License v. 3.0</a>.
		 </p>




	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
