<!DOCTYPE html>
<html lang="en-us">

  <head>


	<meta charset="UTF-8">
	<title>Exploring Genomes Using DIGS</title>
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" type="text/css" href="../stylesheets/normalize.css" media="screen">
	<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" href="../stylesheets/stylesheet.css" media="screen">
	<link rel="stylesheet" type="text/css" href="../stylesheets/github-light.css" media="screen">
	
  </head>
  
  <body>
  
    <section class="page-header">


    
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      
      
      <a href="../index.html" class="btn">Home</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>

    </section>

    <section class="main-content">

	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Overview</strong>
	</h3>
	<hr>


	<p>

	The genomes of Metazoan organisms are packed with dense layers of information.
	For this reason, <strong>genome annotation</strong> is a challenging, long-term endeavour.
	</p>

	<p>
    Consequently, most published genome
	sequences are comprised of DNA  that is incompletely understood in terms of its
	evolutionary origins and functional significance.
	</p>

	<p>
    An exciting aspect of this is that 	
	
	Identifying and characterising the features hidden within this ‘<strong>dark genome</strong>’
	can often reveal useful insights into
	the biology and evolution of genomes.

	</p>


	<p>

	These days, a great deal can be discovered simply by examining genomes "<i>in silico</i>"
	- in other words, by analysing genome sequence data using a computer.
	While experimental studies are required to characterise genome
	features at a functional level, comparative sequence analysis
	can provide insights into how genomes have evolved. 
	These insights can inform the development of hypotheses,
	and the design of experiments to test them.

	</p>


		<img src="images/terra-incognita.jpg"alt="Terra incognita"/>




	<p>

	Sequence similarity searches, like those implemented in the  
	<a target="blank" href='https://blast.ncbi.nlm.nih.gov/Blast.cgi'>basic local alignment search tool</a>
	(BLAST) program, are extremely useful devices for investigating genomes <i>in silico</i>.

	</p>

	<p>

	Difficult to overstate just how useful these are.

	</p>


	<p>

	Using similarity search tools and strategically chosen
	reference datasets, poorly mapped parts of the genomic landscape can be explored
	and described in ways that provide insights into genome biology and evolutionary
	history. 





	</p>
<br>
<blockquote>
		<p>
		  <i>"As more scientists come to rely on genome annotation,
              it will become more important for the scientific community as a whole to
              contribute to this continuing process."
          </i>
          - <b>Lincoln Stein</b>
		</p>
</blockquote>
		 
<br>
	<p>

	These searches can be deployed to investigate genome features have not previously been
	annotated, or have only been incompletely characterised.

	</p>

	<p>

	The genome sequences of most species are a largely unexplored wilderness of
	decayed genes, transposons, and non-coding DNA elements.
	These sequences, even when non-functional, contain a wealth of useful biological information.
	Because similarity searches function independently of annotations,
	they are extremely useful for this 'unexplored land'.

	</p>



	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>The genome's buried treasure</strong>
	</h3>
	<hr>


	<p>

	Progress in genomics has underlined the dynamic nature of genomes,
	which can not only diverge through macrogenomic processes such as chromosomal
	rearrangements and karyotype evolution, but also through the activity of
	transposable elements, and the incorporation of horizontally transferred
	genetic material. 

	</p>

	<p>

	Consequently, most published genome sequences are incompletely annotated.
	One of the things that makes genome annotation challenging, is that 
	much of the useful information is obscured, making it difficult to extract. 
	There are a number of processes that contribute to this:

	</p>

	<p>

	(i) <b>Evolutionary dynamism</b>: as genomes evolve, new genes arise through a
	variety of processes, including segmental duplication, retrotransposition, and gene
	fusion. Whereas some genes undergo relatively measured change under natural 
	selection, others evolve according to selective forces that favor diversification. 

	</p>

	<p>

	For example, genes with antiviral functions evolve under powerful selective forces
	(known as Red Queen dynamics) driven by competition with viral parasites.
	The dynamics can strongly favor divergence and diversification. 

	</p>

	<p>

	Whereas much of the evolutionary dynamism of the genomes is driven by competition with
	parasites in the external environment, another driver is the struggle to suppress 
	intragenomic parasites - such as the various transposable elements that infest eukaryotic
	genomes.

	</p>

	<p>

	For example, mammalian genomes contain an army of genes encoding
	KRAB zinc finger proteins (KZFPs), that specifically target transposable
	elements for epigenetic silencing. 

	</p>

	<p>

	The point is this generates a lot of 'churn'.
	A lot of information that is created as part of the story is embedded in decayed
	sequences. Or else it can be hidden in the way that mobile element sequences are
	distributed throughout species genomes. Since there is a temporal dimension to element 
	activity,  there is information there that be part of the unpicking of the puzzle.  

	</p>


	<p>

	Similar processes operate on genomes, which are partly shaped by
	mutational decay of redundant information.) 

	</p>



	<img src="./images/bryce-erosion.jpg"alt="Bryce Canyon"/>



	<p>

	(ii) <b>Erosion</b>: a by-product of evolutionary dynamism is reduncancy, and
	redundant genes decay, being gradually eroded and fragmented by substitutions, insertions,
	and genomic rearrangements. The degenerated molecular  
	sequences of previously functional genes are called ‘pseudogenes’. 

	</p>


<blockquote>
		<p>
		  Bryce Canyon: Geological structures show the impact of erosion over macroevolutionary time scales.
		</p>
</blockquote>
		 
		 
	<p>

	Pseudogenes can contain useful information about the evolution of genes and genomes,
	and for this reason are worth annotating.
	For example, online resources such as
	<a href='http://pseudogene.org'>pseudogene.org</a>
	that focus specifically on pseudogene annotations.

	</p>

	<p>

	Nonetheless, it's probably fair to say that pseudogenes are somewhat down
	the list of priorities when it comes to genome annotation.
	Furthermore, some will inevitably difficult to recognise if they are very old,
	particularly if they bear little similarity to contemporary functional genes.

	</p>


	<p>

	(iii) <b>Overwriting</b>: another way of thinking about the hidden information in the genome is 
	through comparison to  palimpsests - manuscript pages used in archaeology that
	have been written on more than once, with earlier writing incompletely erased.
	Its important not to take this metaphor too far, but there is a similarity in the way
	that information can be layered.

	</p>


	<p>

	Another way of thinking about the hidden information in the genome is 
	through comparison to  palimpsests - manuscript pages used in archaeology that
	have been written on more than once, with earlier writing incompletely erased.

	</p>

	<p>

	Of course, it is important not to take this metaphor too far, but there is a similarity in
	the way that information can be layered.

	</p>

		<img src="./images/palimpsest.jpg"alt="Palimpsest"/>


	<p>

	(ii) <b>Horizontal transfer</b>:

	</p>

	<p>

	Horizontal acquisition of genes from TEs, viruses and bacteria.
	Sequences that have arisen through horizontal transmission have an unpredictable
	distribution across genomes. This is one reason why they are frequently not captured
	by automated annotation. Other factors include the fact that many are highly degraded
	pseudogenes. 

	</p>



    <br>
	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Exploring genomes *in silico*</strong>
	</h3>
	<hr>



	<p>

	There's a lot of interesting and poorly characterised stuff out there in
	public sequence databases.

	</p>

	<p>

	Much like a web search engine, BLAST can be used to monitor and
	explore an unstable data
	source – the expanding pool of molecular
	sequence data represented in genome sequence
	databases. 

	</p>

	<p>

	Indeed, similarity searches can be used to perform the further characterization of
	sequences retrieved via BLAST. For example, BLAST can be used to detect specific sequence
	features at protein or nucleic acid level (e.g. inversions, repeats, conserved protein
	motifs, oligonucleotide primer sites).  BLAST is also a useful tool for performing rapid
	genotyping of sequences, and for generating alignments, particularly when the sequences
	being aligned have heterogeneous structures.

	</p>

	<p>

	I confess to having spent a great deal of time just using BLAST to 
	sift around in molecular sequence databases - probably more time than I care to account
	for, truth be told.
	Personally,  I find I am easily drawn in when these
	kinds of loose, exploratory investigations seem like they are uncovering new and
	exciting information...

	</p>

	<p>

	<img src="./images/hic-sunt-dracones2.jpg"alt="Palimpsest"/>
	Siren song: there are a variety of ways in which similarity search-based investigations can go awry. 

	</p>


    <p><img src="./images/terra-incognita-squares.png" alt="Here be dragons" /></p>



	<p>

	I say 'seem like' because this kind of loose, heuristic approach to surveying
	genomes involves
	exploring a lot of speculative ideas, many of which (if not most) turn out to be
	unproductive flights of fancy. 

	</p>

	<p>

	Beyond the pitfalls of overactive imagination,
	there are more prosaic challenges to overcome. 
	Simply put - there's an awful lot of molecular sequence data out there.
	Screening this efficiently can be less than straightforward.

	</p>

	<p>

	It can be worthwhile overcoming these challenges though, because sometimes similarity
	search-based investigations lead directly to something interesting and informative
	- like an
	<a href='http://www.pnas.org/content/104/15/6261'>unexpected endogenous virus</a>

	</p>

	<p>

	To facilitate efficient heuristic, similarity
	search-based investigations, we developed an approach called
	<strong>database-integrated genome screening</strong>
	(DIGS), in which the output of similarity searches is captured in a
	relational database.

	</p>

	<p>

	In this blog post, I explain the rationale behind the DIGS approach, illustrating 
	how it can be applied with examples from my own research.
	More broadly though, I want to try and illustrate how interesting
	information can be extracted from published sequence data using only simple,
	publicly available tools, and comparative approaches.  

	</p>

	<p>

	Researchers need tools for exploring sequence databases independently of annotations.
	Sequence databases can be explored by using similarity searches to systematically
	screen for sequences of interest. In database-integrated genome-screening (DIGS), 
	the output of sequence similarity search-based genome 'screens' is captured in a
	relational database. This facilitates the implementation of automated screens that
	can be performed on a large scale. In addition, it allows for the interrogation and
	manipulation of output data using structured query language (SQL), and provides all
	the benefits of a relational database management system (RDBMS) with respect to
	features such as data recoverability, multi-user support and network access.

	</p>

	<p>

	It's extremely easy to get going with this, because there are so many excellent
	online tools for the job, and so much publicly available data.
	Obviously, you first need to decide what genomic 'thing' you want to start investigating.

	</p>

	<p>

	In my case, these tend to be endogenous viral elements (EVEs), but it might be a gene
	family, for example, or a particular type of non-coding sequence. 

	</p>

	<p>

	</p>

	<p>

	BLAST searches are targeted against specific databases
	BLAST lets you confine searches to a specific taxonomic group (for example, viruses).
	It also allows you to selectively exclude certain groups.
	Simply by tweaking these parameters it is possible to gain a great 

	</p>

	<p>

	There is a sense of excitement that comes from seeing the evolutionary processed
	laid out - BLAST searches reflect phylogenetic history.
	It is a very tangible way of connecting with the evolutionary history.

	</p>

	<p>

	Sequence data are constantly accumulating
	(in fact, the rate of data accumulation is accelerating exponentially).
	Thus screens frequently warrant repeating, and as databases expand,
	they become increasingly computationally demanding. 

	</p>

	<p>

	In a typical DIGS workflow, a screen is performed, and sequences recovered via screening
	are then further characterized in line with the objectives of the investigation, and the
	framework is updated accordingly. The overall procedure can henceforth be adjusted and
	reiterated as required. A basic DIGS procedure may also be elaborated, both by
	incorporating additional information into ancillary database tables (e.g. describing 
	characteristics of the species whose genomes are being screened, or of the reference
	sequences being used in screening), and by incorporating project specific analyses in
	the characterization step.

	</p>

	<p>

	Beyond surveying distribution and diversity, 
	what else can we do to investigate evolution.
	There are no hard and fast rules, really.
	Its kind of a case of taking cues where possible.

	</p>

	<p>

	To address this need, we developed an approach called database-integrated genome
	screening (DIGS), in which a relational database is used to capture and analyse
	the output of similarity search-based screens. This facilitates the implementation
	of automated screens that can be performed on a large scale, and provides a flexible,
	well-established framework for interrogating and manipulating the data they produce.

	</p>



	<p>

	In database-integrated genome-screening (DIGS), the output of sequence similarity
	search-based genome 'screens' is captured in a relational database.
	This facilitates the implementation of automated screens that can be performed on a
	large scale. In addition, it allows for the interrogation and manipulation of output
	data using structured query language (SQL), and provides all the benefits of a relational
	database management system (RDBMS) with respect to features such as data recoverability,
	multi-user support and network access.

	</p>








	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
