<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS Tool</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
    <section class="page-header">
    
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      
      <a href="../index.html" class="btn">Home</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">


	   <h3>
		 <a id="inSilicoScreening" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Screening genomes <i>in silico</i></strong>
	   </h3>
       <hr>


	   <p>
		 Sequence similarity search tools, such as the basic local alignment search tool (BLAST),
		 detect regions of local similarity between molecular sequences. Similarity search
		 tools are more-or-less indispensable for biological sequence analysis.
		 They can be used to characterize a locus in detail, helping to identify the coordinates
		 of specific sequence features at protein or nucleic acid level (e.g. conserved protein motifs,
		 oligonucleotide primer sites). They can also be used as a kind of ‘search engine’ for retrieving
		 similar (and thus potentially evolutionarily related) sequences from sequence databases.
		 This second functionality is especially useful for comparative, evolutionary studies,
		 and increasingly so given the speed at which sequence data are now accumulating. 
 	   </p>
 	   
	   <p>
		 The basic functions supplied by BLAST can be elaborated into entire investigative strategies
		 for comparative analysis of genes and genomes. 
 	   </p>
 	   
	   <p>
		 This may involve using different combinations of probe sequences and target databases,
		 or linking BLAST searches with other methods of sequence analysis (e.g. phylogenetic or
		 statistical analysis).
 	   </p>
 	   
	   <p>
		 BLAST-based approaches are especially useful when investigating genomic features that are
		 not well annotated in public sequence databases, such as small RNAs, pseudogenes, transposable elements,
		 highly duplicated gene families, and endogenous viral elements (EVEs). More broadly,
		 BLAST searches can form the backbone of heuristic in silico investigations wherein the overall
		 strategy is loosely defined and there is a requirement to be able to proceed through multiple
		 iterations of trial and error, using new information recovered by in each iteration to update
		 and refine the overall strategy.
 	   </p>
 	   
	   <p>
		 Although systematic BLAST ‘screens’ of genome databases are an important component of many
		 comparative genomics investigations, efficiently implementing these procedures and integrating
		 them into bioinformatics workflows can present a technical challenge.
 	   </p>
 	   
	   <p>
		 The database-integrated genome-screening (DIGS) tool, which aims to provide a robust
		 and extensible framework for implementing systematic, BLAST-based in silico screens
		 of molecular sequence databases and interrogating the data they produce.
		 To demonstrate how the DIGS tool can be applied, we perform a range of  validate the
		 DIGS tool a ‘gold standard’ dataset, and demonstrate its application to an investigation
		 of EVE diversity in mammalian genomes.
 	   </p>


	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
