<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS screening database schema</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
  
  
    <section class="page-header">
    
      <h1 class="project-name">DIGS Tool</h1>
      <h2 class="project-tagline">A software framework for implementing Database-Integrated Genome Screening (DIGS).</h2>
      
      <a href="../../index.html" class="btn">Home</a>
      <a href="./explore.html" class="btn">Background</a>
      <a href="./overview.html" class="btn">Overview</a> 

      <a href="./user-guide.html" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a target="_blank" href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">




	<h3>
	  <a id="dbSchema" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>DIGS project database schema</strong>
	</h3>
	<hr>



    <p>

	  The DIGS <b>project database schema</b> consists of five tables.
	  Of these, only <b>three core tables</b> are relevant to the typical DIGS user,
	  the remaining two (indicated in grey) are used internally by the DIGS tool
	  and can be ignored under ordinary circumstances. 
	  
    </p>


    <p><img src="../assets/images/digs-project-db-schema.jpg" alt="DIGS project DB schema" /></p>



    <p>
      Screening entails performing multiple similarity searches, each involving
      a different <b>probe-target pair</b>. 
      The ’<b>searches performed</b>’ table is used to keep track of
      progress in screening.
       
    </p>



    <p>
      
      The ‘<b>digs_results</b>’ table
	  contains a non-redundant set of all hits obtained in the screen.
       
    </p>



    <p>
  
	  The DIGS tool can be used to execute a ‘<b>consolidation</b>’ process
	  that generates a ‘<b>loci</b>’ table that comprises non-overlapping sets of
	  contiguous matches from the ‘digs_results’ table, based on predefined rules. 

    </p>

    <p>

	  DIGS screening databases can be extended with <b>additional tables</b> linked to the
	  core schema via specific fields.
	  This allows SQL queries to draw on a potentially
	  unlimited range of linked ‘ <b>side-data </b>’.
	  
    </p>

	<br>

	<h3>
	  <a id="fieldDefinitions" class="anchor" href="#fieldDefinitions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Field definitions</strong>
	</h3>
	<hr>

	<p><b>Searches_performed table</b></p>
	
	
	<table>
	
		<thead>
			<tr>
				<th>Field name</th>
				<th>Field type</th>
				<th>Description</th>
			</tr>
		</thead>
	
		<tbody>

			<tr>
				<td><i>probe_lineage</i></td>
				<td>VARCHAR</td>
				<td>Name of the lineage the probe derives from</td>
			</tr>

			<tr>
				<td><i>probe_feature_name</i></td>
				<td>VARCHAR</td>
				<td>Name of the genome feature the probe represents</td>
			</tr>

			<tr>
				<td><i>organism</i></td>
				<td>VARCHAR</td>
				<td>Name of the organism represented by the target database
				(Latin binomial recommended)</td>

			</tr>

			<tr>
				<td><i>target_data_type</i></td>
				<td>VARCHAR</td>
				<td>Target data type (e.g. wgs, transcriptome)</td>
			</tr>


			<tr>
				<td><i>target_version</i></td>
				<td>VARCHAR</td>
				<td>Name/ID of the genome assembly version</td>
			</tr>

			<tr>
				<td><i>target_name</i></td>
				<td>VARCHAR</td>
				<td>Name of the target data file</td>
			</tr>

		</tbody>
		
	</table>

	<br>




	 <p><b>Digs_results & loci tables fields</b></p>
	
	
	 <table>
	
		 <thead>
			 <tr>
				 <th>Name</th>
				 <th>Type</th>
				 <th>Description</th>
			 </tr>
		 </thead>
	
		 <tbody>

			 <tr>
				 <td><i>organism</i></td>
				 <td>VARCHAR</td>
				 <td>Organism name (Latin binomial recommended)</td>
			 </tr>


			 <tr>
				 <td><i>target_data_type</i></td>
				 <td>VARCHAR</td>
				 <td>Target data type (e.g. WGS, transcriptome)</td>
			 </tr>


			 <tr>
				 <td><i>target_version</i></td>
				 <td>VARCHAR</td>
				 <td>Genome build version details</td>
			 </tr>

			 <tr>
				 <td><i>target_name</i></td>
				 <td>VARCHAR</td>
				 <td>Name of the target file</td>
			 </tr>
			
			 <tr>
				 <td><i>scaffold</i></td>
				 <td>VARCHAR</td>
				 <td>Name/ID of the scaffold in which the hit(s) were identified</td>
			 </tr>


			 <tr>
				 <td><i>extract_start</i></td>
				 <td>INT</td>
				 <td><b>5’ (start)</b> position of BLAST hit within the scaffold</td>
			 </tr>

			 <tr>
				 <td><i>extract_end</i></td>
				 <td>INT</td>
				 <td><b>3’ (end)</b> position of BLAST hit within the scaffold</td>
			 </tr>

			 <tr>
				 <td><i>orientation</i></td>
				 <td>VARCHAR</td>
				 <td>Orientation of the BLAST hit within the scaffold</td>
			 </tr>

			 <tr>
				 <td><i>sequence_length</i></td>
				 <td>INT</td>
				 <td>Length of the BLAST hit alignment</td>
			 </tr>

			 <tr>
				 <td><i>sequence</i></td>
				 <td>TEXT</td>
				 <td>DNA sequence of the hit</td>
			 </tr>


			 <tr>
				 <td><i>reference_lineage</i></td>
				 <td>VARCHAR</td>
				 <td>Name of the lineage the probe derives from</td>
			 </tr>

			 <tr>
				 <td><i>reference_feature_name</i></td>
				 <td>VARCHAR</td>
				 <td>Name of the genome feature the probe represents</td>
			 </tr>

			 <tr>
				 <td><i>query_start</i></td>
				 <td>INT</td>
				 <td><b>5’ (start)</b> position of BLAST hit within the scaffold</td>
			 </tr>

			 <tr>
				 <td><i>query_end</i></td>
				 <td>INT</td>
				 <td><b>3’ (start)</b> position of BLAST hit within the scaffold</td>
			 </tr>

			 <tr>
				 <td><i>bitscore</i></td>
				 <td>INT</td>
				 <td>Bitscore of the best match to the hit in the reference library</td>
			 </tr>

			 <tr>
				 <td><i>identity</i></td>
				 <td>VARCHAR</td>
				 <td>Percentage identity of hit to closest matching reference sequence</td>
			 </tr>

			 <tr>
				 <td><i>evalue_num</i></td>
				 <td>INT</td>
				 <td>BLAST 'expect' value numerator</td>
			 </tr>

			 <tr>
				 <td><i>evalue_exp</i></td>
				 <td>INT</td>
				 <td>BLAST 'expect' value exponent</td>
			 </tr>

			 <tr>
				 <td><i>align_len</i></td>
				 <td>INT</td>
				 <td>Length of alignment between hit and best-matching reference sequence</td>
			 </tr>

			 <tr>
				 <td><i>gap_openings</i></td>
				 <td>INT</td>
				 <td>Number of gaps in alignment between hit and best-matching reference sequence</td>
			 </tr>

			 <tr>
				 <td><i>subject_start</i></td>
				 <td>INT</td>
				 <td><b>5’ (start)</b> position of alignment (reference)</td>
			 </tr>
			
			 <tr>
				 <td><i>subject_end</i></td>
				 <td>INT</td>
				 <td><b>3’ (end)</b> position of alignment (reference)</td>
			 </tr>
			
			 <tr>
				 <td><i>query_start</i></td>
				 <td>INT</td>
				 <td><b>5’ (start)</b> position of alignment (extracted hit)</td>
			 </tr>
			
			 <tr>
				 <td><i>query_end</i></td>
				 <td>INT</td>
				 <td><b>3’ (end)</b>  position of alignment (extracted hit)</td>
			 </tr>
			
			 <tr>
				 <td><i>locus_structure</i><br>(loci table only)</td>
				 <td>VARCHAR</td>
				 <td>Captures the order of contiguous hits from digs_results table that were 
				 merged (if multiple hits were merged) to create table row.</td>
			 </tr>

		 </tbody>
		
	 </table>
	 <br>
	 <br>


	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>



    </section>


  </body>
  
</html>
