<!DOCTYPE html>
<html lang="en-us">

  <head>


	<meta charset="UTF-8">
	<title>Exploring Genomes Using DIGS</title>
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
	<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet.css" media="screen">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">
	
  </head>
  
  <body>
  
    <section class="page-header">


    
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      
      
      <a href="../../index.html" class="btn">Home</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a target="blank" href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>

    </section>

    <section class="main-content">



	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Exploring uncharted genomic territory using similarity search tools</strong>
	</h3>
	<hr>


<blockquote>
		<p>
		  <i>"As more scientists come to rely on genome annotation,
              it will become more important for the scientific community as a whole to
              contribute to this continuing process."
          </i>
          - <b>Lincoln Stein</b> (2001)
		</p>
</blockquote>

	

	<p>
	  The genomes of metazoan organisms are packed with dense layers of information,
	  and <strong>genome annotation</strong> is a challenging, long-term endeavour.
	  Consequently, most published genome sequences are comprised of DNA that is
	  incompletely understood in terms of its evolutionary origins and functional significance.
	</p>



	<img src="../assets/images/terra-incognita.jpg"alt="Terra incognita"/>




	<p>
	  Excitingly, much can now be discovered
	  by exploring this 'uncharted genomic territory' "<i>in silico</i>"
	  - in other words, by analysing genome sequence data using a computer.
	  While experimental studies are required to characterise genome
	  features at a functional level, comparative sequence analysis
	  can often provide crucial insights. 
	</p>


	<p>
	  <b>Sequence similarity searches</b>, such as those implemented in the  
	  <a target="blank" href='https://blast.ncbi.nlm.nih.gov/Blast.cgi'><b>basic local alignment search tool (BLAST)</b></a> 
	  program, can be powerful devices for investigating genome sequence data
	  without relying on previously generated annotations. These searches can 
	  be used to recover new information from genomes - especially the 'hidden'
	  information that is not picked up by automated genome annotation tools.	</p>


		 
    <br>

	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Buried treasure - how genomic information can be obscured</strong>
	</h3>
	<hr>


	<p>
	  There are a number of phenomena that complicate and obscure the 
	  evolutionary relationships between genes and other genome features.
	  These include:

	</p>

	<p>

	  (i) <b>Evolutionary 'dynamism'</b>. Whereas some genes undergo relatively
	  measured change under natural selection, others evolve according to selective forces
	  that strongly favor diversification, or are relatively ephemeral.
	  These kinds of selective pressures may generate dynamic and unpredictable
	  patterns of evolution
	  (e.g. rapid change at specific loci, changes in gene structure and copy number) 
	  with the result that evolutionary history is - to a greater or lesser extent -
	  obscured when examining contemporary sequences. 
	</p>


	<p>
	  (ii) <b>Erosion/Decay</b>. 

	  Genome features that become redundant for one reason or another will tend to decay,
	  being gradually eroded and fragmented by mutation.
	</p>



	<img src="../assets/images/bryce-erosion.jpg"alt="Bryce Canyon"/>

<blockquote>
	  Bryce Canyon: Geological structures show the impact of erosion over macroevolutionary time scales.	  
	  Similar processes operate in genomes, which are in part shaped by mutational decay
	  of redundant information. 
</blockquote>
		 
		 
    <br>


		<img src="../assets/images/palimpsest.jpg"alt="Palimpsest"/>


<blockquote>
	  Palimpsests - manuscript pages used in archaeology that
	  have been written on more than once, with earlier writing incompletely erased.
	  Overwriting can also occur at a genetic level, such that older information is obscured
	  but not completely lost. Comparative sequence analysis can sometimes reveal
	  these kinds of phenomena. 
</blockquote>
		 

		 
    <br>


	<p>

	  (iii) <b>Horizontal transfer</b>.
	  Occasionally, metazoan species acquire genes from one another, or from microbes
	  (e.g. viruses and bacteria).
	  Sequences that have arisen through horizontal transmission have an unpredictable
	  distribution across genomes, and are often highly degraded pseudogenes.
	  For these reasons, they are frequently absent from or incompletely captured by
	  existing annotations.

	</p>



		<img src="../assets/images/tangled-tree.jpg"alt="Palimpsest"/>


<blockquote>
	  Tangled trees - horizontal gene transfer creates complex web-like relationships
	  between the genomes of different species, and makes the distribution of related
	  genes somewhat unpredictable. Focussed investigations are often needed to identify 
	  horizontally transferred genes and map their relationships.
</blockquote>
		 

 
	<p>

	  (iv) <b>Transposition</b>. Transposable elements are DNA sequences that are capable
	  of replicating their sequences within species genomes, using either 'copy-and-paste'
	  or 'cut-and-paste' mechanisms.
	  The activity of TEs has profoundly shaped metazoan genome evolution and the attendant
	  genomic 'churn' has generated immense complexity.

	</p>

	<br>



	<p>

	  Because genomes can evolve in a very dynamic way, reconstructing their
	  evolutionary history can be hugely challenging, but also presents an interesting puzzle
	  that can potentially be solved through the diligent use of comparative approaches.
	  These kinds of investigations often rely heavily on <b>similarity search tools</b>.

	</p>



    <br>
	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Exploring genomes <i>in silico</i> using sequence similarity search tools</strong>
	</h3>
	<hr>


	<p>

      Many biologist are 
	  familiar with similarity search tools - specifically <b>BLAST</b>- but typically only use them on
	  an <i>ad hoc</i> or confirmatory basis - for example asking:
	  

	<p>
	  
	  "<b><i>Is the DNA/protein sequence in this text file what I think it is?</b></i>"
	</p>

	<p>
      or
	</p>
	<p>
	  "<b><i>Who else has published the sequence of this gene?</b></i>"
	</p>

	<p>
	  However, similarity search tools are far more powerful than this.
	  They are useful in many different scenarios: for example they can be used to:
	  (i) quickly generate pairwise and multiple sequence 
	  alignments, (ii) perform rapid 'genotyping' of query sequences (e.g. assigning
	  viral sequences to substrains), and (iii) help define the genetic structure of genome features
	  (e.g. by identifying features such as inversions, repeats and conserved domains).  
	</p>

	<p>

      But above and beyond all of this, similarity searches can be used to
      <b>explore the continually expanding pool of
      molecular sequence data</b> represented in genome sequence databases. 

	</p>

	<br>


    <p><img src="../assets/images/terra-incognita-squares.png" alt="Here be dragons" /></p>



<blockquote>
	  The complex world of genomes is only now beginning to be revealed. Go explore!
</blockquote>
		 

	<br>



	<p>
      For small-scale, similarity search-based investigations of this nature, 
      it may be sufficient to simply
      run <i>ad hoc</i> BLAST searches online via a web browser. However, larger-scale projects
      will typically require a more systematic approach, wherein 
      locally held sequence data are systematically 'screened' using a well-defined set of
      query/probe and reference sequences. Since this will often entail large numbers of
      searches, some degree of automation is typically required.

	</p>


	<p>
      Unfortunately, however, such automated screening procedures can be difficult to implement and manage.
      Firstly, exploring genomes <i>in silico</i> is often a heuristic, iterative process,
      and as such requires a degree of oversight.
      This is a significant challenge because screening can often generate large volumes of
      output data that are difficult to organise
      and interrogate without an appropriate analytical framework.
      An additional complication is that the pool of publicly available data is continually
      being updated, making it difficult to keep track and maintain up-to-date information.
      
	</p>


	<p>

      To address these issues, we developed an approach called
      <b>database-integrated genome screening (DIGS)</b>, in which a screening pipeline is
      linked to a <b>relational database management system (RDBMS)</b>, and the outputs are captured
      in a project-specific relational database.
      
      This, database-integrated approach not only provides a convenient and robust basis
      for implementing automated screens that proceed in an efficient, non-redundant way,
      it also allows screening data to be interrogated using <b>structured query language (SQL)</b>
      - a well-established, powerful approach for querying relational databases.
	</p>

	<p>
      Database integration also provides all the advantages of an RDBMS
      with respect to features such as <b>data recoverability</b>, <b>multi-user support</b>,
      and <b>networked data access</b>.
	</p>

 

	<br>
	<h3>
	  <a id="Publications" class="anchor" href="#Publications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Related Publications</strong>
	</h3>
	<hr>

	<p>

	  Zhu H, Dennis T, Hughes J, and RJ Gifford
	  <b>(2018)</b>
	  <br>
	  Database-integrated genome screening (DIGS): exploring genomes heuristically using sequence similarity search tools and a relational database.
	  <a href="https://www.biorxiv.org/content/early/2018/04/25/246835" target="_blank">[preprint]</a> 
	  <br>
	  <br>


	</p>





	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
