<!DOCTYPE html>
<html lang="en-us">

  <head>


	<meta charset="UTF-8">
	<title>Exploring Genomes Using DIGS</title>
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
	<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet.css" media="screen">
	<link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">
	
  </head>
  
  <body>
  
    <section class="page-header">


    
      <h1 class="project-name">The DIGS Tool</h1>
      <h2 class="project-tagline">Explore genomes interactively using BLAST and a relational database.</h2>
      
      
      <a href="../../index.html" class="btn">Home</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/wiki/Installation-and-Setup" class="btn">User Guide</a>
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a target="blank" href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>

    </section>

    <section class="main-content">



	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Exploring uncharted genomic territory using similarity search tools</strong>
	</h3>
	<hr>



		<img src="../assets/images/terra-incognita.jpg"alt="Terra incognita"/>



<blockquote>
		<p>
		  <i>"As more scientists come to rely on genome annotation,
              it will become more important for the scientific community as a whole to
              contribute to this continuing process."
          </i>
          - <b>Lincoln Stein</b> (2001)
		</p>
</blockquote>

	

	<p>
	  The genomes of metazoan organisms are packed with dense layers of information,
	  and <strong>genome annotation</strong> is a challenging, long-term endeavour.
	  Consequently, most published genome sequences are comprised of DNA that is
	  incompletely understood in terms of its evolutionary origins and functional significance.
	</p>



	<p>
	  <b>Sequence similarity searches</b>, such as those implemented in the  
	  <a target="blank" href='https://blast.ncbi.nlm.nih.gov/Blast.cgi'>basic local alignment search tool</a>
	  (BLAST) program, can be powerful devices for investigating genome sequence data
	  without relying on previously generated annotations.
	</p>



	<p>
	  Excitingly, much can now be discovered
	  by examining genomes "<i>in silico</i>"
	  - in other words, by analysing genome sequence data using a computer.
	  While experimental studies are required to characterise genome
	  features at a functional level, comparative sequence analysis
	  can often provide crucial insights. 
	</p>

		 
<br>

	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>The genome's buried treasure</strong>
	</h3>
	<hr>


	<p>
	  As genomes evolve over time, some of the information they contain becomes 
	  obscured. There are a number of processes that contribute to this:

	</p>

	<p>

	  (i) <b>Evolutionary 'dynamism'</b>. Whereas some genes undergo relatively
	  measured change under natural selection, others evolve according to selective forces
	  that strongly favor diversification, or are relatively ephemeral.
	  These more 'chaotic' selective pressures may drive highly dynamic patterns of evolution
	  (e.g. changes in structure and copy number) are to - to a greater or lesser extent -
	  obscured when examining contemporary sequences. 
	</p>


	<img src="../assets/images/bryce-erosion.jpg"alt="Bryce Canyon"/>

<blockquote>
	  Bryce Canyon: Geological structures show the impact of erosion over macroevolutionary time scales.	  
	  Similar processes operate in genomes, which are in part shaped by mutational decay
	  of redundant information. 
</blockquote>
		 

	<p>
	  (ii) <b>Erosion</b>. 

	  Genes that become redundant for one reason or another will tend to decay,
	  being gradually eroded and fragmented by mutation.


      Progress in genomics has underlined the dynamic nature of genomes,
	  which can not only diverge through macrogenomic processes such as chromosomal
	  rearrangements and karyotype evolution, but also through the activity of
	  transposable elements, and the incorporation of horizontally transferred
	  genetic material. 
	  

	</p>




		<img src="../assets/images/palimpsest.jpg"alt="Palimpsest"/>


<blockquote>
	  Palimpsests - manuscript pages used in archaeology that
	  have been written on more than once, with earlier writing incompletely erased.
</blockquote>
		 



	<p>

	  (iii) <b>Horizontal transfer</b>.
	  Occasionally, metazoan species acquire genes from one another, or from microbes
	  (e.g. viruses and bacteria).
	  Sequences that have arisen through horizontal transmission have an unpredictable
	  distribution across genomes, and are often highly degraded pseudogenes.
	  For these reasons, they are frequently absent from or incompletely captured by
	  existing annotations.

	</p>



		<img src="../assets/images/tangled-tree.jpg"alt="Palimpsest"/>



    <br>
	<h3>
		<a id="header-2" class="anchor" href="#header-2" aria-hidden="true"><span class="octicon octicon-link"></span></a><strong>Exploring genomes *in silico*</strong>
	</h3>
	<hr>


	<p>

	  Much like a web search engine, BLAST can be used to monitor and
	  explore an unstable data
	  source â€“ the expanding pool of molecular
	  sequence data represented in genome sequence
	  databases. 

	</p>

	<p>

	  Indeed, similarity searches can be used to perform the further characterization of
	  sequences retrieved via BLAST. For example, BLAST can be used to detect specific sequence
	  features at protein or nucleic acid level (e.g. inversions, repeats, conserved protein
	  motifs, oligonucleotide primer sites).  BLAST is also a useful tool for performing rapid
	  genotyping of sequences, and for generating alignments, particularly when the sequences
	  being aligned have heterogeneous structures.

	</p>


    <p><img src="../assets/images/terra-incognita-squares.png" alt="Here be dragons" /></p>



	<p>

	To facilitate efficient heuristic, similarity
	search-based investigations, we developed an approach called
	<strong>database-integrated genome screening</strong>
	(DIGS), in which the output of similarity searches is captured in a
	relational database.

	</p>



	   <br>
	   <h3>
		 <a id="Publications" class="anchor" href="#Publications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Related Publications</strong>
	   </h3>
	   <hr>

	   <p>

		 Zhu H, Dennis T, Hughes J, and RJ Gifford
		 <b>(2018)</b>
		 <br>
		 Database-integrated genome screening (DIGS): exploring genomes heuristically using sequence similarity search tools and a relational database.
		 <a href="https://www.biorxiv.org/content/early/2018/04/25/246835" target="_blank">[preprint]</a> 
		 <br>
		 <br>


	   </p>





	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
