<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS tool: User Guide</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
    <section class="page-header">
    
      <h1 class="project-name">DIGS Tool</h1>
      <h2 class="project-tagline">A software framework for implementing Database-Integrated Genome Screening (DIGS).</h2>
      
      <a href="../../index.html" class="btn">Home</a>
      <a href="./explore.html" class="btn">Background</a>
      <a href="./overview.html" class="btn">Overview</a>
      <a href="./db-schema.html" class="btn">DB Schema</a>   
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a target="_blank" href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">



	   <h3>
		 <a id="Setup" class="anchor" href="#Setup" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Setting up the DIGS tool environment</strong>
	   </h3>
       <hr>



	   <p>
		  		  
		  Before running DIGS tool, it is necessary to:
 
		 <ol>
		   <b>
			 <li> Install software requirements</li>
			 <li> Set environment variables</li>
			 <li> Set up target libraries</li>
			 <li> Create project-specific sequence libraries</li>
			 <li> Create a  project-specific 'control file'</li>
		   </b>
		 </ol>

 	   </p>



	   <p>
         The sections below provide information about each step.            
 	   </p>




       <br>
	   <h3>
		 <a id="installSoftware" class="anchor" href="#environmentVariables" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>1. Software requirements</strong>
	   </h3>
       <hr>
       

	   <p>
         <u><b>Scripting language</b></u>:<br>
         The DIGS tool requires <a target="_blank" href="https://www.perl.org/"><b>PERL</b></a>.
         The PERL <a target="_blank" href="http://dbi.Perl.org/)"><b>DBI (database interface)</b></a>
         module should be installed via CPAN if it is not part of your native PERL distribution. 
	   </p>



	   <p>
         <u><b>Similarity searches</b></u>:<br>
         The DIGS tool uses the freely available
         <a target="_blank" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi"><b>BLAST+</b></a>
         package to perform sequence similarity searches.
	   </p>


 	   <p>
        
         <u><b>Relational database management system (RDBMS)</b></u>:<br>
         The DIGS tool uses the <b>MySQL</b> relational database management system.
         <a target="_blank" href="http://dev.mysql.com/downloads/mysql/"><b>MySQL Community Server</b></a>
         is a free to use version of this program.
	   </p>

	   

	   <p>
         <u><b>DIGS tool</b></u>:<br>
         The latest version can be obtained <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master"><b>here</b></a>.
         It can be installed on <b>LINUX</b> or <b>UNIX</b> systems.
         Accordingly, installation on Macintosh OSX and Windows 10 and above should be possible, but I have not investigated this personally.
	   </p>



       <br>
	   <h3>
		 <a id="environmentVariables" class="anchor" href="#environmentVariables" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>2. Setting up your local environment for DIGS</strong>
	   </h3>
       <hr>
       
       
       
	   <p>
          The DIGS tool requires users to set four environment variables in their home environment:
          
       <br>
	   <table>
	
		   <thead>
			   <tr>
				   <th>Name</th>
				   <th>Definition</th>
			   </tr>
		   </thead>
	
		   <tbody>

			   <tr>
				   <td><i>$DIGS_GENOMES</i></td>
				   <td>Path to the top-level of the 'target-databases' directory (see below)</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_HOME</i></td>
				   <td>Path to the DIGS-Tool directory</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_MYSQL_USER</i></td>
				   <td>Your mysql user name</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_MYSQL_PASSWORD</i></td>
				   <td>Your mysql password</td>
			   </tr>

		   </tbody>
		
	   </table>
      <br>

   
	   </p>


	   <p>
         Note that one of the environment variables is the path to the 'target-databases' directory.
         DIGS requires that all target sequence databases relevant to the investigation being
         performed are contained within this folder.
         The target genomes directory has a simple, pre-defined subdirectory structure
         (see section immediately below this one).
	   </p>


	   <p>
		 Once environment variables have been set, you should be able to run the DIGS tool
		 script with -h (help) option as shown here: 
		 
  <pre>
  <code>giff01r@Alpha:~/DIGS/DIGS-tool$ ./digs_tool.pl -h'</code></pre>	


   
	   </p>


	   <p>
		 This should print the DIGS input help page to the console: 

  <pre>
  <code>### DIGS version 1.13.2
  
	 ### usage: ./digs_tool.pl m=[option] -i=[control file] -h=[help]

	 ### Main functions

	 -m=1  Prepare target files (index files for BLAST)
	 -m=2  Do DIGS
	 -m=3  Reassign loci
	 -m=4  Defragment loci
	 -m=5  Consolidate loci

	 ### Summarising target databases

	 -g=1  Summarise targets (brief summary, by species)
	 -g=2  Summarise targets (long, by individual target file)

	 ### Managing DIGS screening DBs

	 -d=1  Import tab-delimited data
	 -d=2  Flush core tables
	 -d=3  Drop tables
	 -d=4  Drop a screening DB
	 -d=5  Append data to 'digs_results' table
	 -d=6  Extract sequences using tabular file

	 Target path variable '$DIGS_GENOMES' is set to '/home2/db/digs_genomes'</code></pre>		 

  		 </p>

	   
       <br>
	   <h3>
		 <a id="setUpTargetLibrary" class="anchor" href="#setUpTargetLibrary" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>3. Set up target libraries</strong>
	   </h3>
       <hr>


	   <p>

         The 'target database directory' should have five levels, as shown below.
  	   </p>


       <p><img src="../assets/images/target-directory.jpg" alt="DIGS target directory structure" /></p>

	   <p>
         The top directory level should be the path specified under the DIGS environment
         variable ($DIGS_GENOMES). The top two directory levels can be named anything. 

         The bottom three levels should contain directories with names pertaining to the
         data type (i.e. low coverage, assembly), version, and organism name (Latin binomial
         with underscore – e.g. Homo_sapiens), of the sequence data files they contain*.
 	   </p>



	   <p>


         In the above example, the file 'chrX.fa' would be under the path:

  <pre>
  <code>$DIGS_GENOMES/Mammals/Homo_sapiens/Complete/ncbi_37.3_june_11/ChrX.fa</code></pre>	

 	   </p>


	   <p>
	     Target databases can be indexed for BLAST searching using the DIGS tool,
         by executing as follows:


  <pre>
  <code>./digs_tool.pl –m=1</code></pre>

 	   </p>


	   <p>
         This will initiate a console-based interactive process in which the target 
         database folder is scanned, and genomes that require indexing for BLAST are identified.
 	   </p>


       <br>
	   <h3>
		 <a id="setUpReferenceLibrary" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>4. Set up project-specific sequence librarie</strong>
	   </h3>
       <hr>


	   <p>
          DIGS requires a library of FASTA-formatted sequences, called the 
          'reference sequence library'. This library is used a source of 'probes' for screening, 
          and also provides a means of classifying hits (i.e. similar sequences) identified in these screens.   
 	   </p>



       <p><img src="../assets/images/digs-sequence-format.jpg" alt="DIGS target directory structure" /></p>


	   <p>	   
          The DIGS tool uses a simple rule to capture data from the headers of
          FASTA-formatted reference (and probe) sequences, in which
          everything to the left of the last underscore in the FASTA header
          is taken as the 'species name' and everything to the left is taken 
          as the 'genome feature name'.
 	   </p>


	   <p>
          A subset of reference sequences should be selected as probes.
          The entire reference library can be used - in which case there is no need to
          create a separate file – but it is often sufficient to use only a subset of
          sequences from the reference sequence library, in which case, a separate file
          containing this subset should be created.
          The path to this file is specified in the DIGS control file.
 	   </p>

       <br>
	   <h3>
		 <a id="SetUpControlFile" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>5. Create a project-specific control file</strong>
	   </h3>
       <hr>

	   <p>
          The DIGS control file is a text file that specifies parameters and paths for DIGS.
 	   </p>

	   <p>
         Control files are structured as NEXUS-style blocks delineated by BEGIN and
          ENDBLOCK tokens as shown below.
		  
  <pre>
  <code>Begin SCREENDB;
  db_name=erv_lenti;
  mysql_server=localhost;
  ENDBLOCK;

  BEGIN SCREENSETS;
	  query_aa_fasta=/home/rob/DIGS/projects/lenti-probes.DIGS.faa;
	  reference_aa_fasta=/home/rob/DIGS/projects/ERV-reference.DIGS.faa;
	  bitscore_min_tblastn=60;
	  consolidated_reference_aa_fasta=/home/rob/DIGS/projects/;
	  output_path=./tmp/;
	  seq_length_minimum=50;
	  defragment_range=10;

	  #query_na_fasta=/home/rob/DIGS/projects/lenti-probes.fna
	  #reference_na_fasta=/home/rob/DIGS/projects/lenti-probes.fna
	  #bitscore_min_blastn=30;
  ENDBLOCK;

  BEGIN TARGETS;
	  Mammalia/
  ENDBLOCK;
  </code></pre>

 	   </p>


	   <p>

          Note that not all parameters will need to be defined for every screen.
 
 	   </p>


	

       <br>


	   <h4>
		 <a id="SetUpControlFile" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong><u>Table: Parameters defined in the DIGS control file</u></strong>
	   </h4>
	   <table>
	
		   <thead>
			   <tr>
				   <th>Parameter</th>
				   <th>Definition</th>
			   </tr>
		   </thead>
	
		   <tbody>

			   <tr>
				   <td><i>db_name</i></td>
				   <td>Name of the project-specific screening database</td>
			   </tr>

			   <tr>
				   <td><i>mysql_server</i></td>
				   <td>name of the mysql_server to use (set to 'localhost' to use local database) </td>
			   </tr>

			   <tr>
				   <td><i>query_aa_fasta</i></td>
				   <td>Path to file with amino acid probe sequences</td>
			   </tr>

			   <tr>
				   <td><i>reference_aa_fasta</i></td>
				   <td>Path to file with amino acid reference sequences</td>
			   </tr>

			   <tr>
				   <td><i>query_na_fasta</i></td>
				   <td>Path to file with nucleic acid probe sequences</td>
			   </tr>

			   <tr>
				   <td><i>reference_na_fasta</i></td>
				   <td>Path to file with nucleic acid reference sequences</td>
			   </tr>

			   <tr>
				   <td><i>bitscore_min_tblastn</i></td>
				   <td>Minimum bit-score of tBLASTn hit to extract</td>
			   </tr>

			   <tr>
				   <td><i>bitscore_min_blastn</i></td>
				   <td>PMinimum bit-score of BLASTn hit to extract</td>
			   </tr>

			   <tr>
				   <td><i>seq_length_minimum</i></td>
				   <td>Minimum length of sequence to extract</td>
			   </tr>

			   <tr>
				   <td><i>defragment_range</i></td>
				   <td>Range within which two BLAST hits in the target sequence will be merged</td>
			   </tr>


		   </tbody>
		
	   </table>



       <br>
	   <h3>
		 <a id="PerformingScreening" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Performing DIGS and working with project databases</strong>
	   </h3>
       <hr>



	   <p>

          This section covers a range of different aspects of the database-integrated
          screening (DIGS) process:            


 	   </p>



	   <p>

		 <ol>
		   <b>
			 <li> Running an <i>in silico</i> screen</li>
			 <li> Investigating screening results via a GUI SQL client</li>
			 <li> Updating the reference sequence library & reclassifying data in the 'digs_results' table</li>
			 <li> Incorporating additional tables into a DIGS project database</li>
			 <li> 'Consolidation': Merging & reclassifying hits in the DIGS results table.</li>
		   </b>
		 </ol>


 	   </p>



       <br>
	   <h3>
		 <a id="RunScreen" class="anchor" href="#RunScreen" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>1. Running an <i>in silico</i> screen</strong>
	   </h3>
       <hr>


	   <p>
          Run the DIGS tool as follows.
 	   
 
   <pre>
   <code>./digs_tool.pl -m=2 -i input.ctl</code></pre>

   
	   </p>


	   <p>
	     This will initiate the screening process.
	     A progress log is reported to the screen: 
	   </p>



	   <p>

   <pre>
   <code>

	  Connecting to DB:  eve_1_parvoviridae
	  Created report directory
	  Path: './tmp/result_set_38612_1618435290'
	  Probe sequences:   22 amino acid FASTA sequences
	  Reference library: 39902 amino acid sequences
	  Targets:           1866 target files
	  Previous queries:  41052 previous queries
	  Skipped in set:    41008 (of 41052)
	  Searches to run    44 
	   ### Starting database-integrated genome screening

	  tblastn: 1: 'Eurynorhynchus_pygmeus' (GCA_003697955, low_coverage)
	  target: 'GCA_003697955.1_ASM369795v1_genomic.fna'
	  probe:  'NC_001401-AAV2_NS'
	  
		 # 2 matches to probe: NC_001401-AAV2, NS
		 # 0 matches above threshold (excluded: 0 < length; 2 < bitscore)
		 # done 1 of 44 queries (%2.27)

	  tblastn: 2: 'Hypophthalmichthys_nobilis' (HypNob1.0, low_coverage)
	  target: 'GCA_004193235.1_HypNob1.0_genomic.fna'
	  probe:  'NC_001401-AAV2_NS'

 </code></pre>
   
	   </p>



	   <p>
	     The first few lines of output report the properties of the probe, reference
	     and target sequence libraries that compose the screen. 
	   </p>

	   <p>
	     In addition, the status of the screen is reported - of all the individual BLAST
	     searches that comprise the screen, how many have been performed, and how many are
	     still outstanding. 
	   </p>




	   <p>
	     Run with the -v (verbose) option for more granular logging.
	   </p>




       <br>
	   <h3>
		 <a id="InspectResults" class="anchor" href="#InspectResults" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>2. Investigating screening results using an SQL client</strong>
	   </h3>
       <hr>


	   <p>

          The relational database component allows efficient
          monitoring and summarising the output of screening. 
          Particularly when the screening project database is enriched with additional
          data (e.g. taxonomic data, see below) this greatly
          enhances users capacity to interrogate the data generated by screening.


 	   </p>

	   <p>
          A MySQL client with a graphical user interface (e.g. <b>SequelPro</b>)
          can be used to connect to the screening database.
 	   </p>



	   <p>
		  The following examples illustrate some of the basic ways in 
		  which the RDBMS can be used to monitor and query screening
		  results.
 	   </p>



	   <p>
		  For example, to get counts of matches to each kind of 
		  sequence in our reference library:
 	   </p>

	   <p>
		  
  <pre>
  <code>SELECT  DISTINCT assigned_name, assigned_gene, COUNT(*) AS Number
  FROM   digs_results
  WHERE bitscore >= 60
  GROUP BY  assigned_name, assigned_gene
  ORDER BY  assigned_name, assigned_gene</code></pre>
 
 
 
 		
 	   </p>


 	   
 	   
       <br>
	   <h3>
		 <a id="UpdateRefSeqsAndReassign" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>3. Updating reference sequences and reclassifying results</strong>
	   </h3>
       <hr>

 		
 	   <p>


 
		  
  <pre>
  <code>./digs_tool.pl -m=3 -i ../projects/eve/eve_1_parvo.ctl 
     

	  Connecting to DB:  erv_lenti
	  Created report directory
	  Path: './tmp/result_set_33636_1618435124'
	  Reference library: 39902 amino acid sequences


      Enter a WHERE statement to limit reaasign (Optional) :
      </code></pre>


 	   </p>


 	   


	   <h3>
		 <a id="taxonomyOverview" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>4. Incorporating additional tables into a DIGS project database</strong>
	   </h3>
       <hr>


		 <p>

  To add a table with virus taxonomy:

 <pre>
  <code>giff01r@Alpha:~/DIGS/DIGS-tool$ ./digs_tool.pl -d=1 -i ../projects/eve/eve_1_parvo.ctl 

	######################################################################
	#                                                                    #
	#                   DIGS (version: 1.13.2) 1.13.2                    #
	#                Database-Integrated Genome Screening                #
	#                         Robert J. Gifford                          #
	#                   <robert.gifford@glasgow.ac.uk>                   #
	#                                                                    #
	######################################################################

	 Connecting to DB:  erv_lenti

	 #### WARNING: This function expects a tab-delimited data table with column headers!

	 Please enter the path to the file with the table data and column headings

	 : ../projects/eve/tabular/ncbi_virus_taxonomy.txt</code></pre>		
 	
 	
 	 </p>


 	   
       <br>
	   <h3>
		 <a id="Consolidate" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>5. Merging hits in the DIGS results table into larger sequences</strong>
	   </h3>
       <hr>

	   <p>
         When relying on sequence similarity as a means of recovering the sequences of 
         related genome features, a limitation is that the sequences of many interesting
         genome features are only partially conserved, and large regions of sequence
         within these features may be rearranged or divergent.
 	   </p>

       <p>
         However, when two or more conserved features occur contiguously, their relationship
         can be used to determine the coordinates of a more complete sequence for the 
         genome feature of interest.
 	   </p>



	   <p>
         For example, retroviruses are comprised of three major coding domains
         (<i>gag</i>, <i>pol</i>, <i>env</i> - in that order), flanked by terminal LTRs
         that are usually (though not always entirely) non-coding.
         But while this is the standard form of an integrated retrovirus ('provirus), 
         endogenous retroviruses (ERVs) frequently have much complex genome arrangements,
         with many being fragmentary or mosaic in structure, and large regions of 
         the integrated provirus often being highly divergent from anything seen previously.
 	   </p>



	   <p>
         Accordingly, it makes semse to screen first using individual features 
         (i.e. <i>Gag</i>, <i>Pol</i>, <i>Env</i> polypeptides, plus LTR nucleotide sequences),
         as probes and references, then to '<b>consolidate</b>' the hits to these
         probes into a larger sequences comprised of the hits, plus the intervening sequences.
         At the same time, we can record the relationship between the component parts of
         the merged sequence, where merging occurs.
 	   </p>

	   <p>
         The DIGS tool can be used to implement a ‘consolidation’ of this kind.
         Contiguous hits in the ‘digs_results’ table are merged based on whether they
         are within a user-defined distance of one another.
	   </p>

 


        <p><img src="../assets/images/consolidate.jpg" alt="Overview - phases" /></p>





 
	   <p>
           Running the consolidation process produces a set of merged sequences, and
           also classifies these sequences using the same approach applied when
           generating the digs_results table.
           The results - i.e. a non-overlapping set of sequences, merged as determined by 
           user-specified rules -  are entered into the 'loci' table
           (see the <a href="./db-schema.html>"><b>database schema page</b></a> for details).
           A separate reference sequence library that
           is appropriate for classifying the longer sequences should be used
           for the consolidation, and is
           specified by a distinct parameter (see section 4 in the set-up stages above).
 	   </p>


	   <p>
           The loci table contains most of the same fields as the digs_results table,
           but also includes a 'locus_structure' field that records the relationship
           between merged hits, including their orientation.
 	   </p>


       <br>

       <br>

       <br>



	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
