<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>DIGS tool: User Guide</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="../assets/stylesheets/github-light.css" media="screen">

  </head>
  
  <body>
    <section class="page-header">
    
      <h1 class="project-name">DIGS Tool</h1>
      <h2 class="project-tagline">A software framework for implementing Database-Integrated Genome Screening (DIGS).</h2>
      
      <a href="../../index.html" class="btn">Home</a>
      <a href="./explore.html" class="btn">Background</a>
      <a href="./overview.html" class="btn">Overview</a>
      <a href="./db-schema.html" class="btn">DB Schema</a>   
      <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master" class="btn">Download</a>
      <a target="_blank" href="https://github.com/giffordlabcvr/DIGS-tool" class="btn">View on GitHub</a>
      
    </section>

    <section class="main-content">

 

	   <h3>
		 <a id="Requirements" class="anchor" href="#Requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Requirements</strong>
	   </h3>
       <hr>



	   <p>
         <b>Scripting language</b>:
         The DIGS tool requires <a target="_blank" href="https://www.perl.org/"><b>PERL</b></a>
         with the <a target="_blank" href="http://dbi.Perl.org/)"><b>DBI (database interface)</b></a>
         module installed. 
	   </p>



	   <p>
         <b>Similarity search tools</b>:
         The DIGS tool uses the freely available
         <a target="_blank" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi"><b>BLAST+</b></a>
         package to perform sequence similarity searches.
	   </p>


 	   <p>
        
         <b>Relational database management system (RDBMS)</b>:
         The DIGS tool uses the MySQL RDBMS.
         <a target="_blank" href="http://dev.mysql.com/downloads/mysql/"><b>MySQL Community Server</b></a>
         is a freely downloadable RDBMS that is supported by an active community of open source developers. 
	   </p>

	   

	   <p>
         The DIGS tool can be installed on <b>LINUX</b> or <b>UNIX</b> systems.
         Accordingly, Macintosh installation is also possible. However, due to 
         issues installing PERL's DBI module on Macintosh (discussed
         <a target="_blank" href="http://www.ensembl.info/2013/09/09/installing-perl-dbdmysql-and-ensembl-on-osx/">here</a>,
         <a target="_blank" href="http://www.dimasyusuf.com/installing-dbd-mysql-on-os-x-el-capitan/">here</a>, and
         <a target="_blank" href="https://josephhall.org/nqb2/index.php/dbdmysql_macosx">here</a>),
         I would not recommend attempting a Mac install unless you're familiar with troubleshooting
         these kinds of issues. Similarly, installation on Windows 10 or above via the 
         LINUX subsystem should be possible, but I have not investigated this personally.
	   </p>



	   <!--<img src="../assets/images/user-guide-header.jpg"alt="DIGS User Guide"/>
       <br>-->



       <br>
	   <h3>
		 <a id="Setup" class="anchor" href="#Setup" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Set up the DIGS tool environment</strong>
	   </h3>
       <hr>



	   <p>
         The latest version of the DIGS tool can be obtained <a href="https://github.com/giffordlabcvr/DIGS-tool/zipball/master"><b>here</b></a>
	   </p>


	   <p>
		  
		  
		  Before running DIGS tool, it is necessary to:
 
		 <ol>
		   <li> Set up environment variables required by the DIGS tool</li>
		   <li> Set up target library and index genomes for BLAST</li>
		   <li> Set up reference sequence library and probes</li>
		   <li> Set up a control file</li>
		 </ol>



 	   </p>


	   <h3>
		 <a id="environmentVariables" class="anchor" href="#environmentVariables" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>1. Environment variables</strong>
	   </h3>
       <hr>
	   <p>
          The DIGS tool requires users to set four <b>environment variables</b>:
          



       <br>
	   <table>
	
		   <thead>
			   <tr>
				   <th>Environment variable name</th>
				   <th>Definition</th>
			   </tr>
		   </thead>
	
		   <tbody>

			   <tr>
				   <td><i>$DIGS_GENOMES</i></td>
				   <td>Path to the top-level of the 'target-genomes' directory (see below)</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_HOME</i></td>
				   <td>Path to the DIGS-Tool directory</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_MYSQL_USER</i></td>
				   <td>Your mysql user name</td>
			   </tr>

			   <tr>
				   <td><i>$DIGS_MYSQL_PASSWORD</i></td>
				   <td>Your mysql password</td>
			   </tr>

		   </tbody>
		
	   </table>



       <br>





   
	   </p>


	   <p>
		 Once environment variables have been set, you should be able to run the DIGS tool
		 script: '<b>digs_tool.pl</b>'). This script provides 'one-stop-shopping' for
		 DIGS-related tasks. It can be run in a variety of ways to:
	   </p>



	   <p>

		 <ol>
		   <li> Set-up and run screens</li>
		   <li> Manage project database tables and data</li>
		 </ol>


 	   </p>


	
	   <p>
		 Run with the '-h' option, as shown below, to see options for running DIGS:

  <pre>
  <code>giff01r@Alpha:~/DIGS/DIGS-tool$ ./digs_tool.pl -h



	 ### DIGS version 1.13.2
	 ### usage: ./digs_tool.pl m=[option] -i=[control file] -h=[help]

	 ### Main functions

	 -m=1  Prepare target files (index files for BLAST)
	 -m=2  Do DIGS
	 -m=3  Reassign loci
	 -m=4  Defragment loci
	 -m=5  Consolidate loci

	 ### Summarising target databases

	 -g=1  Summarise targets (brief summary, by species)
	 -g=2  Summarise targets (long, by individual target file)

	 ### Managing DIGS screening DBs

	 -d=1  Import tab-delimited data
	 -d=2  Flush core tables
	 -d=3  Drop tables
	 -d=4  Drop a screening DB
	 -d=5  Append data to 'digs_results' table
	 -d=6  Extract sequences using tabular file

	 Target path variable '$DIGS_GENOMES' is set to '/home2/db/digs_genomes'</code></pre>		 

  		 </p>

	   
       <br>
	   <h3>
		 <a id="setUpTargetLibrary" class="anchor" href="#setUpTargetLibrary" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>2. Set up target library</strong>
	   </h3>
       <hr>

	   <p>
		  The DIGS tool is designed to perform systematic, simialrity-search based screening
		  of <b>local DNA sequence 'databases'</b> - in other words, collections of FASTA
		  formatted DNA sequences.
		  This is the form that most genome and transcriptome assemblies are distributed. 
 	   </p>


       <p><img src="../assets/images/target-directory.jpg" alt="DIGS target directory structure" /></p>


	   <p>
		  The DIGS tool uses the BLAST program suite to perform similarity searches. 
		  BLAST can be used to search any file containing FASTA-formatted sequences for
		  similar sequences. 
 	   </p>


	   <p>

The DIGS tool is designed for use with locally stored, FASTA formatted sequence data. Genome data should be stored in a directory tree with a specific structure. 
 	   </p>



	   <p>

         The 'target genomes directory' should have five levels, as illustrated above.
         The top directory level should be the path specified under the DIGS environment
         variable ($DIGS_GENOMES). The top two directory levels can be named anything. 
  	   </p>



	   <p>

         The bottom three levels should contain directories with names pertaining to the
         data type (i.e. low coverage, assembly), version, and organism name (Latin binomial
         with underscore – e.g. Homo_sapiens), of the sequence data files they contain*.
 	   </p>



	   <p>


In the above example, the file 'chrX.fa' would be under the path:


$DIGS_GENOMES/Mammals/Homo_sapiens/Complete/ncbi_37.3_june_11/ChrX.fa

 	   </p>


	   <p>


         BLAST requires that FASTA files are indexed for similarity searches using the
         'makeblastdb' program that is distributed with the BLAST+ package.
         Because this can be time-consuming when screening many separate files, the digs_tool.pl
         script includes an option to automatically format target FASTA files.
         Note that for this to work properly, FASTA files MUST be labeled with the appropriate
         file extensions (.fa, .fas, or .fasta).

 	   </p>


	   <p>


To run the DIGS tool's genome formatting utility, execute the digs_tool.pl script as follows:


  <pre>
  <code>./digs_tool.pl –m=1</code></pre>		 






 	   </p>




       <br>
	   <h3>
		 <a id="setUpReferenceLibrary" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>3. Set up reference library and probes</strong>
	   </h3>
       <hr>


	   <p>

          DIGS requires a library of FASTA-formatted reference sequences for the gene
          or genetic element under investigation. 
          Probes for screening are selected from the reference library.
          
 	   </p>



       <p><img src="../assets/images/digs-sequence-format.jpg" alt="DIGS target directory structure" /></p>


	   <p>
	   
          The DIGS tool uses a simple rule to capture data from the headers of
          FASTA-formatted reference (and probe) sequences; headers should be structured
          so as to define two hierarchical name elements; ‘name’ and ‘gene_name’,
          separated by an underscore. In the example shown above these are a virus name
          and a gene name.
 	   </p>



	   <p>
          
          Other two-level hierarchical naming schemes (e.g. species &
          gene name, gene-subdomain name) can also be used, providing the same scheme is
          used consistently throughout the project. Reference sequences should be stored
          in a file, the path to which will be specified in the DIGS control file.
          
 	   </p>


	   <p>
	   

          A subset of reference sequences should be selected as probes.
          The entire reference library can be used - in which case there is no need to
          create a separate file – but it is often sufficient to use only a subset of
          sequences from the reference sequence library, in which case, a separate file
          containing this subset should be created.
          The path to this file is specified in the DIGS control file.

 	   </p>



       <br>
	   <h3>
		 <a id="SetUpControlFile" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>4. Set up control file</strong>
	   </h3>
       <hr>



	   <p>
          The DIGS control file contains parameters and paths for DIGS.
 
 	   </p>




	   <p>
         Control files are structured as NEXUS style blocks delineated by BEGIN and
          ENDBLOCK tokens as shown below.
		  
  <pre>
  <code>Begin SCREENDB;
  db_name=erv_lenti;
  mysql_server=localhost;
  ENDBLOCK;

  BEGIN SCREENSETS;
	  query_aa_fasta=/home/rob/DIGS/projects/lenti-probes.DIGS.faa;
	  reference_aa_fasta=/home/rob/DIGS/projects/ERV-reference.DIGS.faa;
	  bitscore_min_tblastn=60;
	  consolidated_reference_aa_fasta=/home/rob/DIGS/projects/;
	  output_path=./tmp/;
	  seq_length_minimum=50;
	  defragment_range=10;

	  #query_na_fasta=/home/rob/DIGS/projects/lenti-probes.fna
	  #reference_na_fasta=/home/rob/DIGS/projects/lenti-probes.fna
	  #bitscore_min_blastn=30;
  ENDBLOCK;

  BEGIN TARGETS;
	  Mammalia/
  ENDBLOCK;
  </code></pre>



 	   </p>


	   <p>

          Note that not all parameters will need to be defined for every screen.
 
 	   </p>



	   <h4>
		 <a id="SetUpControlFile" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Parameters defined in the DIGS control file</strong>
	   </h4>

	

       <br>
	   <table>
	
		   <thead>
			   <tr>
				   <th>Ctl file parameter</th>
				   <th>Definition</th>
			   </tr>
		   </thead>
	
		   <tbody>

			   <tr>
				   <td><i>db_name</i></td>
				   <td>Name of the project-specific screening database</td>
			   </tr>

			   <tr>
				   <td><i>mysql_server</i></td>
				   <td>name of the mysql_server to use (set to 'localhost' to use local database) </td>
			   </tr>

			   <tr>
				   <td><i>query_aa_fasta</i></td>
				   <td>Path to file with amino acid probe sequences</td>
			   </tr>

			   <tr>
				   <td><i>reference_aa_fasta</i></td>
				   <td>Path to file with amino acid reference sequences</td>
			   </tr>

			   <tr>
				   <td><i>query_na_fasta</i></td>
				   <td>Path to file with nucleic acid probe sequences</td>
			   </tr>

			   <tr>
				   <td><i>reference_na_fasta</i></td>
				   <td>Path to file with nucleic acid reference sequences</td>
			   </tr>

			   <tr>
				   <td><i>bitscore_min_tblastn</i></td>
				   <td>Minimum bit-score of tBLASTn hit to extract</td>
			   </tr>

			   <tr>
				   <td><i>bitscore_min_blastn</i></td>
				   <td>PMinimum bit-score of BLASTn hit to extract</td>
			   </tr>

			   <tr>
				   <td><i>seq_length_minimum</i></td>
				   <td>Minimum length of sequence to extract</td>
			   </tr>

			   <tr>
				   <td><i>defragment_range</i></td>
				   <td>Range within which two BLAST hits in the target sequence will be merged</td>
			   </tr>


		   </tbody>
		
	   </table>



       <br>
	   <h3>
		 <a id="PerformingScreening" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Performing DIGS and working with project databases</strong>
	   </h3>
       <hr>




	   <p>

		 <ol>
		   <li> Run a screen and inspect results</li>
		   <li> Updating references & reclassifying data in the 'digs_results' table</li>
		   <li> 'Consolidating' the DIGS results table</li>
		   <li> Adding 'side data' tables</li>
		 </ol>


 	   </p>



       <br>
	   <h3>
		 <a id="RunScreenInspectResults" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>1. Running a screen & examining results</strong>
	   </h3>
       <hr>


	   <p>
          Run the DIGS tool as follows.
 	   
 
   <pre>
   <code>./digs_tool.pl -m=2 -i input.ctl

	  Connecting to DB:  eve_1_parvoviridae
	  Created report directory
	  Path: './tmp/result_set_38612_1618435290'
	  Probe sequences:   22 amino acid FASTA sequences
	  Reference library: 39902 amino acid sequences
	  Targets:           1866 target files
	  Previous queries:  41052 previous queries
	  Skipped in set:    41008 (of 41052)
	  Searches to run    44 </code></pre>

   
	   </p>


	   <p>
          From now, the console will report the results of screening as shown here:
 	   

   <pre>
   <code> ### Starting database-integrated genome screening

	  tblastn: 1: 'Eurynorhynchus_pygmeus' (GCA_003697955, low_coverage)
	  target: 'GCA_003697955.1_ASM369795v1_genomic.fna'
	  probe:  'NC_001401-AAV2_NS'
	  
		 # 2 matches to probe: NC_001401-AAV2, NS
		 # 0 matches above threshold (excluded: 0 < length; 2 < bitscore)
		 # done 1 of 44 queries (%2.27)

	  tblastn: 2: 'Hypophthalmichthys_nobilis' (HypNob1.0, low_coverage)
	  target: 'GCA_004193235.1_HypNob1.0_genomic.fna'
	  probe:  'NC_001401-AAV2_NS'

 </code></pre>
   
	   </p>


	   <p>
          DIGS expedites the interrogation of data derived from BLAST-based screening.
          Data can be interrogated within the framework of the DIGS tool (i.e. by using
          structured query language (SQL)), or exported to external applications and
          examined therein.
 	   </p>


	   <p>
          A MySQL client with a graphical user interface (e.g. <b>SequelPro</b>)
          can be used to connect to the screening database.
 	   </p>



	   <p>
		  Get counts of each.
 	   </p>

	   <p>
		  First download and upack the taxonomy DB from the NCBI FTP server.
		  
  <pre>
  <code>SELECT  DISTINCT assigned_name, assigned_gene, COUNT(*) AS Number
  FROM   digs_results
  WHERE bitscore >= 60
  GROUP BY  assigned_name, assigned_gene
  ORDER BY  assigned_name, assigned_gene</code></pre>
 
 
 
 		
 	   </p>


 	   
 	   
       <br>
	   <h3>
		 <a id="UpdateRefSeqsAndReassign" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>2. Update reference sequences and reassign</strong>
	   </h3>
       <hr>

		  
  <pre>
  <code>./digs_tool.pl -m=3 -i ../projects/eve/eve_1_parvo.ctl 
     

	  Connecting to DB:  erv_lenti
	  Created report directory
	  Path: './tmp/result_set_33636_1618435124'
	  Reference library: 39902 amino acid sequences


      Enter a WHERE statement to limit reaasign (Optional) :
      </code></pre>





 	   
       <br>
	   <h3>
		 <a id="Consolidate" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>3. Consolidate</strong>
	   </h3>
       <hr>

	   <p>

         The DIGS tool can be used to ‘consolidate’ contiguous hits in the ‘digs_results’ table
         according to user-defined
         rules based on:

		 <ul>
		   <li> a minimum length threshold for intervening DNA</li>
		   <li> orientation</li>
		   <li> gene order</li>
		 </ul>



 	   </p>


	   <p>

         
         A fourth database table (“loci”) contains the merged sequences and information about
         their structure.



 	   </p>


	   <h3>
		 <a id="taxonomyOverview" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>4. Using 'side-data' in DIGS databases</strong>
	   </h3>
       <hr>


		 <p>

  To add a table with virus taxonomy:

 <pre>
  <code>giff01r@Alpha:~/DIGS/DIGS-tool$ ./digs_tool.pl -d=1 -i ../projects/eve/eve_1_parvo.ctl 

	######################################################################
	#                                                                    #
	#                   DIGS (version: 1.13.2) 1.13.2                    #
	#                Database-Integrated Genome Screening                #
	#                         Robert J. Gifford                          #
	#                   <robert.gifford@glasgow.ac.uk>                   #
	#                                                                    #
	######################################################################

	 Connecting to DB:  erv_lenti

	 #### WARNING: This function expects a tab-delimited data table with column headers!

	 Please enter the path to the file with the table data and column headings

	 : ../projects/eve/tabular/ncbi_virus_taxonomy.txt</code></pre>		
 	
 	
 	 </p>

       <br>

       <br>

       <br>



	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/DIGS-tool">DIGS for EVEs</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
